// @zoho/lyte-ui-component/mixins/lyte-listview-utils.js
import { Mixin } from "@slyte/core";

class LyteListviewUtilsMixin extends Mixin {
    create_ins() {
		var __this = this,
		$node = __this.$node,
		ns = 'getElementsByClassName',
		elems = $node[ ns ]( 'lyteListviewIntersection' );

		__this._intersectionObs = new window.IntersectionObserver( __this.intersection.bind( __this ), { threshold : [ 0 ], root : $node[ ns ]( 'lyteTableScroll' )[ 0 ] } );

		Array.from( elems ).forEach( this.addToIntersection.bind( this ) );
	}

    destroy_ins() {
		var __this = this,
		ns = '_intersectionObs';

		__this[ ns ].disconnect();
		delete __this[ ns ];
	}

    intersection(elems) {
		var __this = this;

		elems.forEach( function( item ){
			if( item.isIntersecting ){
				var __target = item.target;
				__this.renderNode( __target );
				__this.removeFromIntersection( __target );
			}
		});
	}

    addToIntersection(elem) {
		this._intersectionObs.observe( elem );
	}

    removeFromIntersection(elem) {
		this._intersectionObs.unobserve( elem );
	}

    static observers() {
        return {
            createIntersection : function(){
                this.create_ins();
            }.on( 'didConnect' ),

            destroyIntersection : function(){
                this.destroy_ins();
            }.on( 'didDestroy' )
        };
    }
}

export { LyteListviewUtilsMixin };
