// @zoho/lyte-ui-component/components/javascript/view/listview/filters/lyte-custom-filter.js
import { prop } from "@slyte/core";
import { Component, objectUtils } from "@slyte/component";

class LyteCustomFilterComponent extends Component {
    constructor() {
        super();
    }

    init() {
		this.setData( 'ltPropCondition', { isValid : false, value : "", label : "", type : "custom" } );
	}

    data() {

		return {
			ltPropCondition : prop( 'object' ),
			ltPropReset : prop( 'boolean', { default : false } ),
			ltPropData : prop( 'object', { default : {} } )
		};		
	}

    static observers() {
        return {
            reset_obs : function( arg ){
                if( arg.newValue ){
                    var cb = "onCustomFilterReset",
                    __data = this.data,
                    Lc = objectUtils,
                    condition = __data.ltPropCondition;

                    Lc( condition, 'add', 'isValid', false );
                    Lc( condition, 'add', 'label', "" );
                    Lc( condition, 'add', 'value', "" );

                    this.getMethods( cb ) && this.executeMethod( cb, __data.ltPropData, condition, this.$node );
                    this.setData( arg.item, false );
                }
            }.observes( 'ltPropReset' )
        };
    }
}

export { LyteCustomFilterComponent };
