// @zoho/lyte-ui-component/components/javascript/lyte-board.js
import { prop } from "@slyte/core";
import { Component, LyteUiComponentComponentRegistry } from "../component.js";
import $L from "@zoho/lyte-dom";
/**
 * Renders an board
 * @component lyte-board
 * @version 3.1.0
 * @methods onBoardScroll
 */
class LyteBoardComponent extends Component {
    constructor() {
        super();
    }

    data() {
		return {
		 	/**
            * @componentProperty {object} ltPropBoardDetail
            * @version 3.1.0
            */ 
			lyteViewPort : prop("boolean", {"default" : true}),//No I18n
			'ltPropBoardDetail' : prop( 'object' ), 
			/**
            * @componentProperty {number} ltPropIndex
            * @version 3.1.0
            */ 
			'ltPropIndex' : prop( 'number' ),
			/**
            * @componentProperty {boolean} ltPropBoardSortable=true
            * @version 3.1.0
            */
            'ltPropBoardSortable' : prop( 'boolean', {
				'default':  window._lyteUiUtils.resolveDefaultValue( 'lyte-board', 'boardSortable', true )
			} ),
			/**
            * @componentProperty {string} ltPropClass
            * @version 3.1.0
            */
			'ltPropClass' : prop('string',{
				'default':  window._lyteUiUtils.resolveDefaultValue( 'lyte-board', 'class', '' )
			}),
			/**
            * @componentProperty {string} ltPropKanbanId=''
            * @version 3.1.0
            */
			'ltPropKanbanId' : prop( 'string', {
				'default' : ''
			} ) ,
			/**
            * @componentProperty {boolean} ltPropMoreStageRecord=false
            * @version 3.1.0
            */
			'ltPropMoreStageRecord' : prop( 'boolean', {
				'default': false
			} ),
			/**
            * @componentProperty {string} ltPropNoResultMessage=''
            * @version 3.1.0
            */
			'ltPropNoResultMessage' : prop('string',{'default': window._lyteUiUtils.resolveDefaultValue( 'lyte-dropdown', 'noResultMessage', window._lyteUiUtils.i18n( 'no.results.found' ) )}),
			/**
            * @componentProperty {boolean} ltPropLoadingYield=true
            * @version 3.1.0
            */
			'ltPropLoadingYield' : prop( 'boolean',{
				'default':true
			}),
			/**
            * @componentProperty {string} ltPropCardClassName
            * @version 3.1.0
            */
			'ltPropCardClassName' : prop('string',{'default':undefined}),
			/**
            * @componentProperty {boolean} ltPropAria
            * @version 3.1.0
            */
		   'ltPropAria' : prop('boolean',{
			   'default':false
		   }),
		   /**
            * @componentProperty {object} ltPropBoardAria
            * @version 3.1.0
            */
		   'ltPropBoardAria' : prop('object',{
				default:{}
			}), 
			/**
            * @componentProperty {object} ltPropCardAria
            * @version 3.1.0
            */
			'ltPropCardAria' : prop('object',{
				default:{}
			}), 
			
			'ltPropNoResultYield' : prop('boolean',{
				default:false
			}),
			/**
            * @experimental dummyId
            */
			'dummyId' : prop( 'string', {
				'default' : ''
			} ),
			/**
            * @experimental cardArray
            */
			'cardArray' : prop( 'array', {
				'default':[]
			} )
		};		
	}

    init() {

		this.setCardArray()
	}

    didConnect() {
		this.$node.getVisibleCard =function(){
			    var scrollDiv = this.$node.getElementsByClassName('lyteKanbanNestedSortable')[0]
			    if(scrollDiv.scrollHeight>scrollDiv.clientHeight){
			    	return this.getVisibleNode()

			    }
			return (this.getData('ltPropBoardDetail').cards ? this.getData('ltPropBoardDetail').cards  : [])

		}.bind(this)
		this.$node.collapse = function(){
			var kanbanviewItem = $L(this.$node ).closest( '.lyteKanbanViewItem' )[ 0 ]
			if(kanbanviewItem){
				kanbanviewItem.classList.add( 'lyteKanbanBoardCollapse' )
			} 
			
		}.bind( this )
		this.$node.expand = function(){
			var kanbanviewItem = $L(this.$node ).closest( '.lyteKanbanViewItem' )[ 0 ]
			if(kanbanviewItem){
				kanbanviewItem.classList.remove( 'lyteKanbanBoardCollapse' )
			} 		
		}.bind( this )
	}

    addAria() {
		if(this.getData('ltPropAria') ){
			var self = this
			window._lyteUiUtils.setAttribute( this.$node.querySelector('.lyteBoardWrapper'), this.getData( 'ltPropBoardAria' ) || {}, {} );
			this.$node.querySelectorAll('.lyteBoardItemContentData').forEach(function(item){
				window._lyteUiUtils.setAttribute( item, self.getData( 'ltPropCardAria' ) || {}, {} );
			})

		}
	}

    didDestroy() {
		clearTimeout( this.timeout1 );
		clearTimeout(this.debounceTimeout)
	}

    getVisibleNode() {
        // return;
        var bcr=this.$node.getBoundingClientRect(),
        originalRows = Array.from( this.$node.getElementsByClassName( 'lyteBoardItemContentData' ) ),
        tValue = Math.max( bcr.top + 10 , -10 ),
        bValue = Math.min( window.innerWidth + 10, bcr.bottom),
        visible = [], boardDetails = this.getData('ltPropBoardDetail').cards;
        


       
        for( var i = 0; i < originalRows.length; i++ ){
            var row = originalRows[ i ],
            _bcr = row.getBoundingClientRect();
            if( _bcr.bottom > tValue && _bcr.top <bValue ){
                visible.push( boardDetails[i] );
                
            }
        }

        
        return visible;
    
	}

    setCardArray() {
		var boardDetail = this.getData('ltPropBoardDetail')
		if( boardDetail && boardDetail.cards ){
			this.setData( 'cardArray', boardDetail.cards )
		}
	}

    addSortableForNewRecords() {
		var div = this.$node.querySelectorAll( '.lyteKanbanNestedSortable.'+this.getData('ltPropKanbanId') )[0]
		if(div.classList.contains('sortable-parent')){
			var sortableClass = div.getSortableClass(),
			cardWithoutSortable = this.$node.querySelectorAll( '.lyteBoardItemContentData:not(.'+sortableClass+')' )
			$L(cardWithoutSortable).map(function(index,element){
				element.parentElement.addToSortable(element);
			})
		}
		
		
	}

    addShadow() {
		this.$node.querySelector( '.lyteBoardWrapper ' ).classList.add( 'lyteKanbanviewShadow' ); 
		this.$node.querySelector('.lyteBoardHeader').classList.remove('lyteKanbanviewHeaderShadow'); // No I18n

	}

    hasScrollHeightReached(event) {
		if ( event.target.scrollHeight - 10 <= ( Math.ceil( event.target.offsetHeight ) + Math.ceil( event.target.scrollTop ) ) ) {
			if ( this.getData( 'ltPropMoreStageRecord' ) && this.getMethods( 'onBoardScroll' ) ) {
				this.executeMethod( 'onBoardScroll' , this.getData( 'ltPropBoardDetail' ), this, event  ); //NO i18n
			}
		}
	}

    removeShadow() {
		this.$node.querySelector( '.lyteBoardWrapper ' ).classList.add( 'lyteKanbanviewShadow' ); 
		this.$node.querySelector('.lyteBoardHeader').classList.remove('lyteKanbanviewHeaderShadow'); // No I18n
	}

    executeScrollStop(event) {
		if( this.getMethods( 'onBoardScrollStop' ) ) {
			var visible,boardDetail = this.getData('ltPropBoardDetail'),
			 scrollDiv = this.$node.getElementsByClassName('lyteKanbanNestedSortable')[0]
			if(scrollDiv.scrollHeight>scrollDiv.clientHeight){
			    	visible = this.getVisibleNode()
			 } else{
			 	visible = boardDetail.cards
			 }
			this.executeMethod( 'onBoardScrollStop', boardDetail, visible, this, scrollDiv.scrollTop, event );
		}
	}

    static actions() {
        return {
            boardScroll : function( event ) {
                if ( event.target.scrollTop != 0 ) {
                    this.addShadow();
                }

                this.timeout1 = setTimeout(function() {
                    
                    this.hasScrollHeightReached(event)
                }.bind( this ), 10 );

                if ( event.target.scrollTop == 0 ){
                    this.removeShadow()
                }
                //debounce
                clearTimeout(this.debounceTimeout)
                this.debounceTimeout = setTimeout( function(){
                    this.executeScrollStop(event)
                }.bind(this),100)
                // event.preventDefault()
                // event.stopPropagation();
            }
        };
    }

    static observers() {
        return {
            ariaObs : function(){
                if(!this.getData('ltPropLoadingYield')){
                    this.addAria()
                }
            }.observes('ltPropBoardAria','ltPropCardAria','ltPropLoadingYield').on('didConnect'),

            viewPortObs : function(){
                    if( !this.getData('lyteViewPort') && this.getData('ltPropBoardSortable')) {
                        var kanbanview = $L(this.$node ).closest( 'lyte-kanbanview' )[ 0 ]
                        if(kanbanview){
                            kanbanview.component.addSortableForCard()
                        } 
                        this.addAria()
                    }
                

            }.observes('lyteViewPort'),

            contentObs : function(){
                this.setCardArray()
                if(this.getData('ltPropBoardSortable') && !this.getData('lyteViewPort')){
                    this.addSortableForNewRecords()
                }
                
            }.observes( 'ltPropBoardDetail.cards.[]' )
        };
    }
}

LyteUiComponentComponentRegistry.registerHelper('lyteUiAddSortableClass',function(sortable,item ){
	if(sortable && item.parentElement && item.parentElement.getSortableClass){
		return "sortable-element "+item.parentElement.getSortableClass()+(item.classList.contains("sortable-element-selected") > -1 ? "sortable-element-selected ":" ")
	}
	return ''	
});
export { LyteBoardComponent };
