// @zoho/lyte-ui-component/components/javascript/lyte-timeline.js
import { prop } from "@slyte/core";
import { Component, LyteUiComponentComponentRegistry } from "../component.js";

class LyteTimelineComponent extends Component {
    constructor() {
        super();
    }

    data() {
		return {

			// left, right, alternate
			'ltPropAlignment': prop( 'string', { 'default': 'left' } )
		};		
	}
}

if( !window._lyteUiUtils.registeredCustomElements[ 'lyte-timeline-item' ] ) {
    window._lyteUiUtils.registeredCustomElements[ 'lyte-timeline-item' ] = true;

    /**
 	 * @customElement lyte-timeline-item
 	 */

    class LyteTimelineItemCustomElements extends LyteUiComponentComponentRegistry.CustomElements {
        static options = {clone : {allCallbacks : false}};

        lookups() {
            return [{
                component: LyteUiComponentComponentRegistry
            }];
        }

        constructor() {
            super();
        }

        static get observedAttributes() {
            return [];
        }

        connectedCallback() {
            var lyteNode = this.$node;
            var separator = document.createElement( 'lyte-timeline-separator' ),
			label = lyteNode.querySelector( 'lyte-timeline-item-label' );

            lyteNode.insertBefore( separator, label.nextElementSibling );
        }
    }

    LyteTimelineItemCustomElements.register("lyte-timeline-item");
}
export { LyteTimelineComponent };
