// @zoho/lyte-ui-component/components/javascript/lyte-drawer.js
import { prop } from "@slyte/core";
import { Component } from "../component.js";
import $L from "@zoho/lyte-dom";

window.addCloseEvent = function() {

	document.addEventListener('click',function(event){
		var ele = event.target;
		while(!$L(ele).hasClass('modalWrapper') && ele.tagName != "LYTE-DRAWER-BODY" && ele.tagName != "LYTE-DRAWER" && ele.tagName !="LYTE-DRAWER-FREEZE" && ele.tagName != "LYTE-MODAL-FREEZE" && ele.tagName != 'LYTE-DROP-BOX' && ele.tagName != 'HTML'){
            ele = ele.parentElement;
            if(!ele){
                return
            }
        }
		if(ele.tagName == 'HTML' || ele.tagName == "LYTE-MODAL-FREEZE" || ele.tagName == "LYTE-DRAWER-FREEZE"){
			var last = window._LyteDrawer_.length-1;
			if(last > -1){
				if(window._LyteDrawer_[last].tagName == "LYTE-DRAWER" && window._LyteDrawer_[last].ltProp('show') && window._LyteDrawer_[last].ltProp("overlayClose")){
					if(window._LyteDrawer_[last]){
						window._LyteDrawer_[last].ltProp('show',false);
					}
				}
			}
		}
	},true);
	document.addEventListener('keydown',function(event){
			event = event || window.event;
            var isEscape = false;
            if ("key" in event) {
                isEscape = (event.key == "Escape" || event.key == "Esc");
            } else {
                isEscape = (event.keyCode == 27);
            }
            if (isEscape) {
				var last = window._LyteDrawer_.length-1;
				if(last > -1){
					if(window._LyteDrawer_[last].tagName == "LYTE-DRAWER" && window._LyteDrawer_[last].ltProp('show') && window._LyteDrawer_[last].ltProp("closeOnEscape")){
						if(window._LyteDrawer_[last]){
							window._LyteDrawer_[last].ltProp('show',false);
						}
					}
				}
            }
	},true);
};

/**
 * Renders a Drawer
 * @component lyte-drawer
 * @version  4.0.0
 * @methods onBeforeShow,onShow,onBeforeClose,onClose,onSelected
 */
if(!window._LyteDrawer_){
    window._LyteDrawer_ = [];
}

class LyteDrawerComponent extends Component {
    constructor() {
        super();
    }

    data() {
		return {
			/** 
			 * @componentProperty  {left | right} ltPropPosition=left
			 */
			ltPropPosition:prop("string",{"default":"left"}),
			/** 
			 * @componentProperty  {string} ltPropWidth=200px
			 */
			ltPropWidth:prop("string",{"default":"200px"}),
			/** 
			 * @componentProperty  {string} ltPropHeight=100%
			 */
			ltPropHeight:prop("string",{"default":"100%"}),
			/** 
			 * @componentProperty  {object} ltPropModal={}
			 */
			ltPropModal:prop("object",{"default":{}}),
			/** 
			 * @componentProperty  {boolean} ltPropFreeze=true
			 */
			ltPropFreeze:prop("boolean",{"default":true}),
			/** 
			 * @componentProperty  {string} ltPropAnimationDuration=0.3s
			 */
			ltPropAnimationDuration:prop("string",{"default":"0.3s"}),
			/** 
			 * @componentProperty  {boolean} ltPropMiniVariant=false
			 */
			ltPropMiniVariant:prop("boolean",{"default":false}),
			/** 
			 * @componentProperty  {string} ltPropUserValue=name
			 */
			ltPropUserValue: prop("string",{default:"name"}),
			/** 
			 * @componentProperty  {string} ltPropSystemValue=value
			 */
			ltPropSystemValue:prop("string",{default:"value"}),
			/** 
			 * @componentProperty {array} ltPropOptions
			 * @default []
			 */
			ltPropOptions:prop("array",{default:[]}),
			/** 
			 * @componentProperty  {boolean} ltPropShow=false
			 */
			ltPropShow:prop("boolean",{"default":false}),
			/** 
			 * @componentProperty  {string} ltPropSelectedClass
			 */
			ltPropSelectedClass:prop("string"),
			/** 
			 * @componentProperty  {string} ltPropSelected
			 */
			ltPropSelected:prop("string"),
			/** 
			 * @componentProperty  {boolean} ltPropCloseOnSelect=false
			 */
			ltPropCloseOnSelect:prop("boolean",{"default":false}),
			/** 
			 * @componentProperty  {boolean} ltPropOverlayClose=true
			 */
			ltPropOverlayClose : prop("boolean",{"default":true}),
			/** 
			 * @componentProperty  {boolean} ltPropShowOpenButton=true
			 */
			 ltPropShowOpenButton : prop("boolean",{"default":true}),
			/** 
			 * @componentProperty  {boolean} ltPropShowCloseButton=true
			 */
			 ltPropShowCloseButton : prop("boolean",{"default":true}),
			 /** 
			 * @componentProperty  {boolean} ltPropCloseOnEscape=true
			 */
			ltPropCloseOnEscape : prop("boolean",{"default":true}),
			/** 
			 * @componentProperty  {array} ltPropDisabledList
			 * @default []
			 */
			 ltPropDisabledList:prop("array",{"default":[]}),
			/** 
			 * @componentProperty  {string} ltPropWrapperClass
			 */
			ltPropWrapperClass:prop("string"),
			/** 
			 * @componentProperty  {inline | overlay |  inlineOverlay} ltPropLayout=overlay
			 */
			ltPropLayout: prop("string",{"default":"overlay"}),
			/** 
			 * @componentProperty  {string} ltPropExpandMiniVariant="click"
			 */
			ltPropExpandMiniVariant: prop("string",{
				"default": "click"
			}),
			/** 
			 * @componentProperty  {string} ltPropMiniVariantWidth=50px
			 */
			ltPropMiniVariantWidth : prop("string",{"default": "50px"}),
			/** 
			 * @componentProperty  {boolean} ltPropYield=false
			 */
			ltPropYield:prop("boolean",{default:false}),
			ltPropDataTabindex : prop("string",{"default" : ""}),

			returnedFalse : prop("boolean",{"default" : false}),
			currentPosition : prop("string"),
			config : prop("object",{
				"default" : {}
			}),
			isMouseEvent : prop("boolean",{"default":false}),
			modalAttr : prop("object",{"default": {}})
		};		
	}

    // methods calling start
    selected(selectedValue, lyteDrawerItem) {
		if(this.getMethods("onSelected")){
			this.executeMethod("onSelected",selectedValue,lyteDrawerItem,this); 
		}
	}

    beforeShow(skip) {
		if(!skip && this.getMethods("onBeforeShow")){
			return this.executeMethod("onBeforeShow", this); 
		}
	}

    show(skip) {
		if(!skip && this.getMethods("onShow")){
			this.executeMethod("onShow", this); 
		}
	}

    beforeClose(skip) {
		if(!skip && this.getMethods("onBeforeClose")){
			return this.executeMethod("onBeforeClose", this); 
		}
	}

    close(skip) {
		if(!skip && this.getMethods("onClose")){
			this.executeMethod("onClose", this); 
		}
	}

    // methods calling end

    getDrawerForModal() {
		var actualModalDiv = this.$node.querySelector("lyte-modal").component.actualModalDiv;
		if(actualModalDiv) {
			return actualModalDiv.querySelector("lyte-modal-content");
		}
	}

    getParentElement() {
		/* 
			get parent element based on layout
		*/
		var layout = this.data.ltPropLayout;
		if(layout == "overlay"){
			return this.getDrawerForModal();
		}
		return this.$node.querySelector(".lyteDrawerInlineBody");
	}

    getDrawerPanel() {
		var layout = this.data.ltPropLayout;
		if(layout == "overlay"){
			return this.getDrawerForModal();
		}
		return this.$node.querySelector(".lyteDrawerPanel");
	}

    getLyteDrawerItem(parentElement, drawerItemvalue) {
		/* 
			get lyte drawer item from parent element
		*/
		return parentElement.querySelector("[data-value ='"+window._lyteUiUtils.escape(drawerItemvalue)+"']");
	}

    getActiveDrawerItem(parentElement) {
		/* 
			get active drawer item using active class
		*/
		return parentElement.querySelector(".lyteDrawerActiveItem");
	}

    activeClassActionForDrawerItem(element, action) {
		/*
		function used to add or remove activeClass from a element
		*/ 
		if(element){
			var selectedClass = this.data.ltPropSelectedClass;
			element.classList[action]("lyteDrawerActiveItem");
			if(selectedClass){
				element.classList[action](selectedClass);
			}
		}
	}

    removePreviouslySelected(parent) {
		/* 
			removing the selected class from the previously selected element
		*/
		var prevActiveDrawerItem = this.getActiveDrawerItem(parent);
		this.activeClassActionForDrawerItem(prevActiveDrawerItem, "remove");
	}

    addDisableClassForDrawerItem(parentElement, value) {
		/* 
			adding disabled class to drawerItem
		*/
		var element = this.getLyteDrawerItem(parentElement, value);
		if(element){
			element.classList.add("lyteDrawerDisabledItem");
		}
	}

    removeDisableClassFromDrawerItem(element) {
		/* 
			remove disabled class from lyte drawer item
		*/
		element.classList.add("lyteDrawerDisabledItem");
	}

    selectDrawerItem() {
		/* 
			selecting drawerItem using ltPropSelected
		*/
		var close = this.data.ltPropCloseOnSelect,
		parent = this.getDrawerPanel();
		if(parent){
			var selected = this.data.ltPropSelected;
			var curActiveDrawerItem = this.getLyteDrawerItem(parent, selected);
			this.removePreviouslySelected(parent);
			this.activeClassActionForDrawerItem(curActiveDrawerItem, "add");
			if(curActiveDrawerItem){
				if(close){
					this.setData("ltPropShow",false);
				}
			}
		}
		else {
			this.setData("config.selection",true);
		}
	}

    disableDrawerItems() {
		/* 
			disabling drawerItem using ltPropDisableList
		*/
		var array = this.getData("ltPropDisabledList"),
		parent = this.getDrawerPanel();
		if(parent){
			var disabledlist = parent.querySelectorAll(".lyteDrawerDisabledItem");
			for(var index = 0 ; index<disabledlist.length;index++){
				this.removeDisableClassFromDrawerItem(disabledlist[index]);
			}
			for(var index = 0 ; index<array.length;index++){
				this.addDisableClassForDrawerItem(parent, array[index]);
			}
		}
		else {
			this.setData("config.disable",true);
		}
	}

    removeDrawerFromStore() {
		/* 
		removing the lyteDrawer node from global variable(_LyteDrawer_)
		*/
		var drawer = this.$node;
		var lyteDrawers =  _LyteDrawer_;
		var drawerlength = lyteDrawers.length;
		for(var index=0; index<drawerlength; index++) {
			if(lyteDrawers[index] === drawer) {
				lyteDrawers.splice(index,1);
				break;
			}
		}
	}

    getPosition() {
		/* checking the rtl direction and getting correct position */
		return this.data.ltPropPosition;
	}

    initializeDataForModal() {
		/* setting initial data for modal like slideFromLeft or slideFromRight*/
		var lyteModal = this.$node.querySelector("lyte-modal");
		var offset = lyteModal.ltProp("offset"),
		currentPosition = this.getPosition(),
		seconds =  (this.data.ltPropAnimationDuration || '').replace(/s/g, '');
		lyteModal.ltProp("transition",{
			"animation": "slideFrom"+window._lyteUiUtils.capitalize(currentPosition),
			"duration":seconds
		});
		lyteModal.ltProp("offset",{[currentPosition] : "0px", top: offset.top});
	}

    getFreezeLayer() {
		/* To get freeze layer element from dom*/
		return this.$node.querySelector("lyte-drawer-freeze");
	}

    showFreezeLayer() {
		/* To show freeze layer for inline and inlineOverlay drawer*/
		if(this.data.ltPropFreeze) {
			var freezeLayer = this.getFreezeLayer();
			freezeLayer.style.display = "block";
			setTimeout(function(){
				freezeLayer.classList.add('lyteDrawerFreezeLayerShown');
			},0);
		}
	}

    hideFreezeLayer(initialStage) {
		/* To hide freeze layer for inline and inlineOverlay drawers*/
		if(this.data.ltPropFreeze) {
			var freezeLayer = this.getFreezeLayer();
			freezeLayer.classList.remove('lyteDrawerFreezeLayerShown');
		}
	}

    focusDrawerPanel() {
		this.getDrawerPanel().focus();
	}

    showInlineDrawer() {
		/* Used to open inline and inlineOverlay drawer*/
		var returnValue =  this.beforeShow();
		if(returnValue === false){
			this.setData({
				"returnedFalse": true,
				"ltPropShow": false
			});
		}
		else{
			this.showFreezeLayer();
			this.show();
			_LyteDrawer_.push(this.$node);
		}
	}

    closeInlineDrawer() {
		/* close the opened inline and inlineOverlay drawer*/
		var returnValue = this.beforeClose();
		if(returnValue === false){
			this.setData({
				"returnedFalse": true,
				"ltPropShow": true
			});
		}
		else{
			this.hideFreezeLayer();
			this.close();
			this.removeDrawerFromStore();	
		}
	}

    isFocusedLyteDrawerItem(element) {
		var focusClassName = "lyteDrawerItemFocused";
		return element.classList.contains(focusClassName);
	}

    isActiveLyteDrawerItem(element) {
		var activeClassName = "lyteDrawerActiveItem";
		return element.classList.contains(activeClassName);
	}

    isDisabledLyteDrawerItem(element) {
		var disabledClassname = "lyteDrawerDisabledItem";
		return element.classList.contains(disabledClassname);
	}

    isLyteDrawerItem(element) {
		if(element && element.tagName == "LYTE-DRAWER-ITEM" && 
		!this.isDisabledLyteDrawerItem(element)) {
			return true;
		}
	}

    getActiveOrFocusedItem(elements) {
		var len = elements.length,
		activeIndex;
		for(var index=0; index<len; index++) {
			if(this.isFocusedLyteDrawerItem(elements[index])){
				return index;
			}
			else if(this.isActiveLyteDrawerItem(elements[index])) {
				activeIndex =  index;
			}
		}
		return activeIndex;
	}

    isNotActiveElement(element) {
		if(element) {
			return !element.classList.contains("lyteDrawerActiveItem");
		}
	}

    isValidFocusableItem(element) {
		return !element.classList.contains("lyteDrawerDisabledItem");
	}

    findNextActive(elements, index, forward) {
		var increment = forward ? 1 : -1,
		eleLen = elements.length;
		if(index === undefined){
			index = forward ? 0 : eleLen-1;
		}
		else {
			index = index + increment;
		}
		for( ;  forward ? index < eleLen : index > -1; index = index + increment ) {
			if( this.isValidFocusableItem(elements[index])) {
				return elements[index];
			}
		}
	}

    getAllDrawerItems() {
		var parentElement = this.getDrawerPanel();
		return parentElement.querySelectorAll("lyte-drawer-item");
	}

    isValidateElement(element) {
		return element != document 
		&& element != document.body
		&& element != document.documentElement 
		 && element.tagName != 'LYTE-DRAWER-BODY';
	}

    elementsFromPoint(x, y) {
        var elements = [], 
		element = document.elementFromPoint(x, y), 
		prevElement;
        while (this.isValidateElement(element)) {
            element._pointerEvents = element.style.pointerEvents;
            element.style.pointerEvents = 'none';
            elements.push(element);
            prevElement = element;
            element = document.elementFromPoint(x, y);
            if (prevElement === element) {
                break
            }
        }
        for (var i = 0; i < elements.length; i++) {
            elements[i].style.pointerEvents = elements[i]._pointerEvents;
            delete elements[i]._pointerEvents;
        }
        return elements;
    }

    openDrawer() {
		var parentElement = this.getDrawerPanel();
		var expandMiniVariant =  this.data.ltPropExpandMiniVariant;
		var eventName = this.data.isMouseEvent?"mouseenter":expandMiniVariant;
		this.setData("ltPropShow",true);
		if(this.data.ltPropShow) {
			parentElement.removeEventListener( eventName, this._bindedDrawerOpen, true);
			if(this.data.isMouseEvent) {
				document.addEventListener( "mousemove", this._bindedDrawerClose, true);
			}
		}
	}

    closeOnHover(event) {
        if (this.$node) {
            var elements = document.elementsFromPoint ? document.elementsFromPoint(event.clientX, event.clientY) 
			: this.elementsFromPoint(event.clientX, event.clientY);
			var parentElement = this.getDrawerPanel();
            if (elements.indexOf(parentElement) == -1 ) {
				this.setData("ltPropShow", false);
				if(!this.data.ltPropShow) {
					document.removeEventListener( "mousemove", this._bindedDrawerClose, true);
					parentElement.addEventListener( "mouseenter", this._bindedDrawerOpen, true);
				}
            }
        }
    }

    isMiniVariant() {
		return this.data.ltPropMiniVariant;
	}

    isMouseEvent(eventName) {
		return /^(mouseenter|mousemove|mouseover|hover)$/.test(eventName);
	}

    initializeEvent() {
		if(this.isMiniVariant()) {
			var parentElement = this.getDrawerPanel();
			if(parentElement) {
				var expandMiniVariant = this.data.ltPropExpandMiniVariant;
				var eventName = expandMiniVariant;
				if(eventName) {
					var isMouseEvent =  this.isMouseEvent(eventName);
					this._bindedDrawerOpen = this.openDrawer.bind(this);
					if(isMouseEvent) {
						this._bindedDrawerClose = this.closeOnHover.bind(this);
						eventName = "mouseenter";
					}
					parentElement.addEventListener(eventName,this._bindedDrawerOpen,true);
					this.setData("isMouseEvent", isMouseEvent);
				}
			}
		}
	}

    checkAndUpdateSelectedValue(config) {
		if(config.selection) {
			this.selectDrawerItem();
			this.$addon.objectUtils(config, "delete", "selection");
		}
	}

    checkAndUpdateDisbaleList(config) {
		if(config.disable) {
			this.disableDrawerItems();
			this.$addon.objectUtils(config, "delete", "disable");
		}
	}

    initializeDrawerDefaultData() {
		var config = this.data.config;
		this.checkAndUpdateSelectedValue(config);
		this.checkAndUpdateDisbaleList(config);
	}

    moveIntoView(element) {
		var panel = this.getDrawerPanel(),
		panelScrollTop = panel.scrollTop,
		elementTop = element.offsetTop;

		if( elementTop <= panelScrollTop ) {
			panel.scrollTop = elementTop;
		}
		else {
			panel.scrollTop =elementTop + element.offsetHeight - panel.offsetHeight;
		}
	}

    updateModalAttr() {
		var modalAttr =  this.$addon.deepCopyObject(this.data.ltPropModal);
		var notAllowed = ["width", "height","showCloseButton","closeOnEscape", 
		"allowMultiple", "overlayClose","freeze", "wrapperClass"];
		notAllowed.forEach(function(item){
			if(modalAttr.hasOwnProperty(item)) {
				delete modalAttr[item]
			}
		});
		this.setData("modalAttr",  modalAttr);
	}

    removeAttachedEvents() {
		if(this.isMiniVariant()){
			var parentElement = this.getDrawerPanel();
			var	expandMiniVariant = this.data.ltPropExpandMiniVariant;
			if(expandMiniVariant) {
				var eventName = this.data.isMouseEvent?"mouseenter":expandMiniVariant; 
				if(this.data.isMouseEvent){
					document.removeEventListener( "mousemove", this._bindedDrawerClose, true);
				}
				parentElement.addEventListener( eventName, this._bindedDrawerOpen, true);
				delete this._bindedDrawerClose;
				delete this._bindedDrawerOpen;
			}
		}
	}

    initializeShow() {
		var show  = this.data.ltPropShow;
		if(show) {
			var layout = this.data.ltPropLayout;
			if(layout === "inline" ||  layout === "inlineOverlay"){
				this.showInlineDrawer(!window.changes);
			}
		}
	}

    removeMiniVariantEvent() {
		var parentElement = this.getDrawerPanel();
		var eventName =  this.data.ltPropExpandMiniVariant;
		if(!this.isMouseEvent(eventName)) {
			parentElement.addEventListener(eventName,this._bindedDrawerOpen,true);
		}
	}

    // lifecycle hooks start

    init() {
		this.initializeShow();
	}

    didConnect() {
		this.initializeEvent();
	}

    didDestroy() {
		//remove documnet event listener
		this.removeDrawerFromStore();
		this.removeAttachedEvents();
	}

    //observers end

    static actions() {
        return {

            selectedItem:function(event,type){
                var closestItem = $L(event.target).closest("lyte-drawer-item", this.$node)[0];
                if(this.isLyteDrawerItem(closestItem)) {
                    var value = closestItem.getAttribute("data-value");
                    this.setData("ltPropSelected",value);
                    this.selected( this.data.ltPropSelected, closestItem);
                }
            },

            makeDrawerItemActive : function(event) {
                
                var keyCode = event.keyCode;
                var drawerItems = this.getAllDrawerItems();
                var activeIndex = this.getActiveOrFocusedItem(drawerItems);
                var activeItem =  drawerItems[activeIndex];
                if(this.data.ltPropShow){
                    if( keyCode === 38  || keyCode === 40 ) {
                        var forward = keyCode === 40;
                        var nextActiveItem = this.findNextActive(drawerItems, activeIndex, forward);
                        if(nextActiveItem) {
                            if(activeItem){
                                activeItem.classList.remove("lyteDrawerItemFocused");
                            }
                            nextActiveItem.classList.add("lyteDrawerItemFocused");
                            this.moveIntoView(nextActiveItem);
                            event.preventDefault();
                        }
                    }
                    else if(keyCode === 13 && this.isNotActiveElement(activeItem)) {
                        var value = activeItem.getAttribute("data-value");
                        activeItem.classList.remove("lyteDrawerItemFocused");
                        this.setData("ltPropSelected",value);
                        this.selected( this.data.ltPropSelected, activeItem);
                        event.preventDefault();
                    }
                }

            },

            selectActiveItem : function(event){
                var closestItem = $L(event.target).closest("lyte-drawer-item", this.$node)[0];
                var drawerPanel = this.getDrawerPanel();
                var focusedItem = drawerPanel.querySelector(".lyteDrawerItemFocused");
                if(focusedItem) {
                    focusedItem.classList.remove("lyteDrawerItemFocused");
                }
                if(closestItem) {
                    closestItem.classList.add("lyteDrawerItemFocused");
                }
            },

            openDrawer : function() {
                this.setData("ltPropShow",true);
            },
            
            closeDrawer : function() {
                this.setData("ltPropShow",false);
            },

            updateFreezeLayerStyle : function(node) {
                if(this.data.ltPropShow ===  false) {
                    node.style.setProperty("display", "none");
                }
            }

        };
    }

    static methods() {
        return {
            /* These below methods will work for overlay drawer only*/

            modalOnBeforeShow:function(component){
                this.initializeDataForModal(component.$node);
                return this.beforeShow();
            },

            modalOnShow:function(component){
                this.initializeDrawerDefaultData();
                this.show();
                this.focusDrawerPanel();
                _LyteDrawer_.push(this.$node);
            },

            modalOnBeforeClose:function(event,component){
                return this.beforeClose();
            },

            modalOnClose:function(component){
                if(this.isMiniVariant()) {
                    return;
                }
                this.removeDrawerFromStore();
                this.close();
            }

        };
    }

    static observers() {
        return {
            // lifecycle hooks end

            //observers start
            selectedObserver : function(changes){
                this.selectDrawerItem();
            }.observes("ltPropSelected").on('didConnect'),

            disabledListObserver : function(){
                this.disableDrawerItems();
            }.observes("ltPropDisabledList").on('didConnect'),

            showChanges:function(changes){
                if(this.data.returnedFalse){
                    this.setData('returnedFalse',false);
                    return;
                }
                var  show  = this.data.ltPropShow,
                layout = this.$node.ltProp("layout"),
                position = this.getPosition();
                this.setData("currentPosition",position);
                if(show){
                    if(layout === "inline" ||  layout === "inlineOverlay"){
                        this.showInlineDrawer(!changes);
                    }
                }
                else {
                    if(layout == "inline" || layout === "inlineOverlay"){
                        this.closeInlineDrawer(!changes);
                        if(this.isMiniVariant()) {
                            this.removeMiniVariantEvent();
                        }
                    }
                }
            }.observes("ltPropShow").on('didConnect'),

            styleObserver : function(observerChange){
                // all animation are handled using css variable in css
                if(this.data.ltPropLayout != "overlay"){
                    var drawerPanel =  this.getDrawerPanel();
                    var compData =  this.data;
                    var cssVarMapping  =  {
                        ltPropWidth : "--lyte-drawer-width",
                        ltPropMiniVariantWidth : "--lyte-drawer-mini-variant-width",
                        ltPropAnimationDuration : "--lyte-drawer-transition-duration"
                    }
                    if(observerChange) {
                        var key = observerChange.item;
                        drawerPanel.style.setProperty( cssVarMapping[key], compData[key]);
                    }
                    else {
                        for(var key in cssVarMapping) {
                            drawerPanel.style.setProperty( cssVarMapping[key], compData[key]);
                        }
                    }
                }
            }.observes("ltPropWidth","ltPropAnimationDuration","ltPropMiniVariantWidth").on("didConnect"),

            modalAttrObserver :  function(){
                this.updateModalAttr();
            }.observes("ltPropModal")
        };
    }
}

//need to be moved as common code
if (document.readyState === "complete" || document.readyState === "interactive"){
    window.addCloseEvent();
}
else{
    document.addEventListener("DOMContentLoaded", function(){
        window.addCloseEvent();
    });
}
export { LyteDrawerComponent };
