// @zoho/lyte-ui-component/components/javascript/lyte-radiobutton-group.js
import { prop } from "@slyte/core";
import { Component } from "../component.js";
import $L from "@zoho/lyte-dom";

/**
 * Renders a radiobutton group
 * @component lyte-radiobutton-group
 * @version 3.22.0
 * @methods onBeforeChecked,onBeforeUnchecked,onChecked,onUnchecked,onChanged
 */
class LyteRadiobuttonGroupComponent extends Component {
    constructor() {
        super();
    }

    data() {
		return {
			/**
			 * @componentProperty {default|primary|secondary|switch|slider} ltPropType
			 * @default default
			 */
			ltPropType : prop("string",{
				"default": window._lyteUiUtils.resolveDefaultValue( 'lyte-radiobutton-group', 'type', 'default' )
			}),
			/**
			 * @componentProperty {string} ltPropName
			 */
			ltPropName : prop("string",{"default":undefined}),
			/** 
			 * @componentProperty {horizontal | vertical} ltPropAlignment=vertical
			 */
			ltPropAlignment : prop("string",{
				"default": window._lyteUiUtils.resolveDefaultValue( 'lyte-radiobutton-group', 'alignment', "Vertical" )
			}),
			/**
			 * @componentProperty {array} ltPropOptions
			 * @default []
			 */

			ltPropOptions : prop("array",{"default":[]}),
			/** 
			 * @componentProperty {object} ltPropSelected={}
			 */
			ltPropSelected : prop("object",{"default":{}}),
			/** 
			 * @componentProperty {string} ltPropLabelClass=""
			 */
			ltPropLabelClass : prop("string",{
				"default":  window._lyteUiUtils.resolveDefaultValue( 'lyte-radiobutton-group', 'labelClass', "" )
			}),
			/** 
			 * @componentProperty {string} ltPropClass=""
			 */
			ltPropClass: prop( 'string', { 
				'default': window._lyteUiUtils.resolveDefaultValue( 'lyte-radiobutton-group', 'class', "" )
			}),
			/** 
			 * @componentProperty {boolean} ltPropFireOnInit=false
			 */
			ltPropFireOnInit : prop("boolean",{
				"default": window._lyteUiUtils.resolveDefaultValue( 'lyte-radiobutton-group', 'fireOnInit', false )
			}),
			/**
			 * @componentProperty {string} ltPropUserValue=name
			 */
			ltPropUserValue : prop("string",{"default":"name"}),
			/**
			 * @componentProperty {string} ltPropSystemValue=value
			 */
			ltPropSystemValue : prop("string",{"default":"value"}),
			/**
			 * @componentProperty {array} ltPropDisabledList
			 * @default []
			 */
			ltPropDisabledList : prop("array",{"default":[]}),
			/** 
			 * @componentProperty {boolean} ltPropFocus=false
			 */
			ltPropFocus : prop("boolean",{"default":false}),
			/** 
			 * @componentProperty {array} ltPropAriaAttributes
			 * @default []
			 */
			ltPropAriaAttributes : prop("array",{"default":window._lyteUiUtils.resolveDefaultValue( 'lyte-radiobutton-group', 'ariaAttributes', [] )}),
			/** 
			 * @componentProperty {boolean} ltPropYield=false
			 */
			ltPropYield : prop("boolean",{"default":false}),
			/** 
			 * @componentProperty {boolean} ltPropPreventFocus=false
			 */
			ltPropPreventFocus : prop("boolean",{"default":false}),
			/** 
			 * @componentProperty {boolean} ltPropReadOnly=false
			 */
			ltPropReadOnly : prop("boolean",{"default":false}),
			/** 
			 * @componentProperty {string} ltPropRadioBtnClass=''
			 */
			ltPropRadioBtnClass : prop("string",{"default":window._lyteUiUtils.resolveDefaultValue( 'lyte-radiobutton-group', 'radioBtnClass', '' )}),
			/** 
			 * @componentProperty {string} ltPropSelectedValue={}
			 */
			ltPropSelectedValue : prop("string",{"default":""}),
			/** 
			 * @componentProperty {string} ltPropSelectedValue=""
			 */
			ltPropSelectedClass :  prop("string",{"default":""}),

			prevSelectedValue : prop("object",{"default":{}}),//used to store previous selected value
			preventFocusSet : prop("boolean",{"default":false}),
			selThroScript : prop("boolean",{"default":false})
			//selThroScript is used to prevent the recursive while setting the ltPropSelected through script.
		};		
	}

    getAllRadiobuttons() {
		//return this.$node.querySelectorAll("lyte-radiobutton");
		return $L("lyte-radiobutton",this.$node);
	}

    makeRadiobuttonFocus() {//need to be tested
		var radiobuttons = this.getAllRadiobuttons();
		for(var iterator=0;iterator<radiobuttons.length;iterator++){
			var radiobutton = radiobuttons[iterator];
			if(!radiobutton.ltProp("disabled")){
				radiobutton.focus();
				break;
			}
		}
	}

    updateDisabledValue(array, value) {
		if(array){
			var comp = this;
			array.forEach(function(item){
				var radiobutton = comp.getRadiobuttonFromSystemValue(item);
				if(radiobutton){
					radiobutton.ltProp("disabled",value);
				}
			});
		}
	}

    radiobuttonStateChange(radiobutton, value) {
		this.setData("selThroScript",true);
		radiobutton.ltProp("checked",value);
		if(this.data.ltPropSelectedClass) {
			var func = value ? "add" : "remove";
			radiobutton.classList[func](this.data.ltPropSelectedClass);
		}
		this.setData("selThroScript",false);
	}

    findSelectedObjectFromOptions(selectedValue) {
		var options = this.data.ltPropOptions;
		var systemValue = this.data.ltPropSystemValue;
		for(var index = 0; index < options.length ; index++) {
			var item = options[index];
			if(item[systemValue] === selectedValue) {
				return item;
			}
		}
	}

    getSelectedValueAsString(selectedObject) {
		var systemValue = this.data.ltPropSystemValue;
		return  selectedObject ? selectedObject[systemValue] : "";
	}

    init() {
		var selectedValue = this.data.ltPropSelectedValue;
		var selected = this.data.ltPropSelected;
		if(selectedValue) {
			var selectedObject = this.findSelectedObjectFromOptions(selectedValue);
			if(selectedObject) {
				this.setData("ltPropSelected", selectedObject);
			}
		}
		else if(selected) {
			this.setData("ltPropSelectedValue", this.getSelectedValueAsString(selected));
		}
	}

    didConnect() {
        var lyteSelf = this;
        this.updateDisabledValue(this.getData("ltPropDisabledList"),true);
        this.preventPointerEvent();
        this.$node.updateAriaArray = function(index, value) {
			var ltPropAriaAttributes = this.component.data.ltPropAriaAttributes;
			lyteSelf.$addon.arrayUtils(ltPropAriaAttributes, "replaceAt", index, value)
		};
		if(this.data.ltPropSelectedValue && this.data.ltPropSelectedClass) {
			var selectedRadioButton =  this.getRadiobuttonFromSystemValue(this.data.ltPropSelectedValue);
			selectedRadioButton && selectedRadioButton.classList.add(this.data.ltPropSelectedClass);
		}
    }

    didDestroy() {
		delete this.$node.updateAriaArray;
	}

    isNotEmpty(object) {
		var systemValue = this.getData("ltPropSystemValue");
		if(object && object[systemValue] !=  undefined){
			return true;
		}
		return false;
	}

    getRadiobutton(radioButtonvalue) {
		var systemValue = this.getData("ltPropSystemValue");
		if(this.isNotEmpty(radioButtonvalue)){
			return $L('[data-value="'+radioButtonvalue[systemValue]+'"]',this.$node)[0];
		} 
	}

    getRadiobuttonFromSystemValue(systemValue) {
		return $L('[data-value="'+systemValue+'"]',this.$node)[0];
	}

    selectedChanges(oldValue, newValue) {
		var radiobutton =  this.getRadiobutton(newValue);
		this.setPreviousValue(oldValue);
		if(radiobutton){
			this.radiobuttonStateChange(radiobutton,true);
		}
		else if(!this.isNotEmpty(newValue)){
			var oldRadiobutton = this.getRadiobutton(oldValue);
			if(oldRadiobutton){
				this.radiobuttonStateChange(oldRadiobutton,false);
			}
			this.setPreviousValue(newValue);
			/* if the newValue is empty then we also making prevSelectedValue as empty 
			because if they again select radiobutton by click, there prevSelectedValue will be different.
			basically we are reseting the prevSelectedValue.
			*/
		}
	}

    preventPointerEvent() {
		var radioBtnGrp = this.$node;
		if(this.data.ltPropReadOnly) {
			radioBtnGrp.classList.add('lyteRadioBtnGroupReadOnly');
		}
		else {
			radioBtnGrp.classList.remove('lyteRadioBtnGroupReadOnly');
		}
	}

    getPreviousValue(bypass) {
		if(this.getData("selThroScript") || !bypass){
			/* Here if the selThroScript is true then the prevSelectedValue is setted 
			through selectedObserver so we can use current prevSelectedValue
			if bypass is true, the prevSelectedValue is not yet set and 
			considering the ltPropSelcted as previous value.
			In all cases,if the selThroScript is true then the value is setted and
			we can get the currentValues of prevSelectedValue and ltPropSelected respectively.
			*/
			return this.getData("prevSelectedValue");
		}
		return this.getSelectedValue();
	}

    setPreviousValue(value) {
		if(value){
			this.setData("prevSelectedValue",value);
		}
		else if(!this.getData("selThroScript")){
			//Here if selThroScript is false, we setting current ltPropSelected as prevSelectedValue.
			//This will be done before updating the current selected value.
			this.setData("prevSelectedValue",this.getSelectedValue());
		}
	}

    getSelectedValue(radiobutton) {
		if(this.getData("selThroScript") || !radiobutton){
			return this.getData("ltPropSelected");
		}
		return this.getCurrentSelectedValue(radiobutton);
	}

    setSelectedValue(radiobutton) {
		if(!this.getData("selThroScript")){
			var oldValue = this.getData("ltPropSelected");
			var newValue = this.getCurrentSelectedValue(radiobutton);
			var systemValue = this.data.ltPropSystemValue;
			if(!oldValue || oldValue[systemValue] != newValue[systemValue]){
				//Above check is used to prevent the from setting the value in ltPropOnInit
				this.setData("selThroScript",true);
				this.setData("ltPropSelected",this.getCurrentSelectedValue(radiobutton));
				this.setData("ltPropSelectedValue", newValue[systemValue]);
				this.setData("selThroScript",false);
			}
		}
	}

    getCurrentSelectedValue(radiobutton) {
		var options = this.getData("ltPropOptions");
		var index  = radiobutton.getAttribute("index");
		return options[index];
	}

    rollbackSelectedValue() {
		if(this.getData("selThroScript")){
			this.setData("ltPropSelected",this.getPreviousValue());
		}
	}

    static methods() {
        return {
            rdbBeforeUnchecked : function( input, component, userAction){
                var previousValue = this.getPreviousValue(true);
                var prevRadiobutton = this.getRadiobutton(previousValue);
                if(this.getMethods('onBeforeUnchecked')){
                    if(this.executeMethod('onBeforeUnchecked', this, previousValue, prevRadiobutton, userAction) === false){
                        this.rollbackSelectedValue();
                        return false;
                    }
                }
            },
            rdbUnchecked : function( input, component, userAction){
                this.setPreviousValue();
                var previousValue = this.getPreviousValue();
                var prevRadiobutton = this.getRadiobutton(previousValue);
                if(this.getMethods('onUnchecked')){
                    this.executeMethod('onUnchecked', this, previousValue, prevRadiobutton, userAction);
                }
            },
            rdbBeforeChecked : function( input, component, userAction){
                if(this.getMethods('onBeforeChecked')){
                    if(this.executeMethod('onBeforeChecked', this, this.getSelectedValue(component.$node), component.$node, userAction) === false){
                        this.rollbackSelectedValue();
                        return false;
                    }
                }
            },
            rdbChecked : function( input, component, userAction){
                var radiobutton = component.$node;
                this.setSelectedValue(radiobutton);
                if(this.getMethods('onChecked')){
                    this.executeMethod('onChecked', this, this.getSelectedValue(), radiobutton, userAction);
                }
            },
            rdbChanged : function( input, component, userAction){
                var prevSelectedValue = this.getPreviousValue(),
                curSelectedValue =  this.getSelectedValue(),
                prevRadiobutton = this.getRadiobutton(prevSelectedValue),
                curRadioButton =  this.getRadiobutton(curSelectedValue);
				if(this.data.ltPropSelectedClass) {
					prevRadiobutton && prevRadiobutton.classList.remove(this.data.ltPropSelectedClass);
					curRadioButton && curRadioButton.classList.add(this.data.ltPropSelectedClass);
				}
                if(this.getMethods('onChanged')){
                    this.executeMethod('onChanged', this, prevSelectedValue, curSelectedValue, prevRadiobutton, curRadioButton, userAction);
                }
            }
        };
    }

    static observers() {
        return {
            selectedObserver : function(changes){
                if(this.getData("selThroScript")){
                    return;
                }
                var newValue = changes.newValue;
                this.selectedChanges(changes.oldValue, newValue);
                this.fromObserver = true;
                this.setData("ltPropSelectedValue", this.getSelectedValueAsString(newValue));
                this.fromObserver = false;
            }.observes("ltPropSelected"),

            selectedValueObserver : function(changes){
                if(this.getData("selThroScript") || this.fromObserver){
                    return;
                }
                var selectedObject = this.findSelectedObjectFromOptions(changes.newValue);
                this.setData("ltPropSelected", selectedObject);
            }.observes("ltPropSelectedValue"),

            disabledValueObserver : function(changes){// need to be tested
                this.updateDisabledValue(changes.oldValue,false);
                this.updateDisabledValue(changes.newValue,true);
            }.observes('ltPropDisabledList'),

            focusObserver : function(){
                var focus = this.getData("ltPropFocus");
                if(focus){
                    this.makeRadiobuttonFocus();
                }
                this.setData("ltPropFocus",false);
            }.observes("ltPropFocus").on('didConnect'),

            readOnlyObserver : function() {
                this.preventPointerEvent();
            }.observes('ltPropReadOnly')
        };
    }
}
//TODO: window._lyteUiUtils.escape() check.
/**
 * @syntax nonYielded
 * <lyte-radiobutton-group lt-prop-user-value="name" lt-prop-system-value="value"  
 * lt-prop-name="group-1" lt-prop-options='[{"name": "IOS","value": "apple"},{"name": "Android","value": "google"},{"name": "Window","value": "mircosoft"}]'>
 * </lyte-radiobutton-group>
 */

/**
 * @syntax yielded
 * <lyte-radiobutton-group  lt-prop-user-value="name" lt-prop-system-value="value"
 * lt-prop-name="group-1" lt-prop-options='[{"name": "IOS","value": "apple"},{"name": "Android","value": "google"},{"name": "Window","value": "mircosoft"}]'>
 *     <template is="registerYield" yield-name="yield">
 *        <span>{{ltItem.name}}</span>
 *     </template>
 * </lyte-radiobutto-group>
 */

export { LyteRadiobuttonGroupComponent };
