// @zoho/lyte-ui-component/components/javascript/lyte-video.js
import { prop } from "@slyte/core";
import $L from "@zoho/lyte-dom";
import { Component, LyteUiComponentComponentRegistry } from "../component.js";

/**
 * This component is used to render a video content in the document
 * @component lyte-video
 * @version 3.45.0
 * @dependency lyte-multislider
 *  components/lyte-multislider
 * 	theme/compiledCSS/default/ltr/lyte-ui-slider.css
 * @methods onVolumeChange, onProgress, onPause, onPlay, onScreenShot, onMenuClick, onBeforePreFetch, onPreFetchSuccess,onPreFetchError, onEnded 
 * @utility play,pause,screenShot 
**/
class LyteVideoComponent extends Component {
    constructor() {
        super();
    }

    data() {
		return {
            /**
			 * @componentProperty {Object[]} ltPropSource
			 * @version 3.45.0
			 */
			ltPropSource : prop( 'array' ),
            /**
			 * @componentProperty {Object[]} ltPropTracks
			 * @version 3.45.0
			 */
			ltPropTracks : prop( 'array' ),
            /**
			 * @componentProperty {number} ltPropVolume=1
			 * @version 3.45.0
			 */
			ltPropVolume : prop( 'number', { default : 1 }),
            /**
			 * @componentProperty {boolean} ltPropMuted
			 * @version 3.45.0
			 */
            ltPropMuted : prop( 'boolean' ),
            /**
			 * @componentProperty {number} ltPropPlayRate=1
			 * @version 3.86.0
			 */
			ltPropPlayRate : prop( 'number', { default : 1 }),
            /**
			 * @componentProperty {string} ltPropCrossOrigin=''
			 * @version 3.45.0
			 */
			ltPropCrossOrigin : prop('string', { default : ''}),
            /**
			 * @componentProperty {number} ltPropCurrentTime
			 * @version 3.45.0
			 */
			ltPropCurrentTime : prop( 'number' ),
            /**
			 * @componentProperty {Object[]} ltPropChapters
			 * @version 3.45.0
			 */
			ltPropChapters : prop( 'array' ),
            /**
			 * @componentProperty {string} ltPropWidth='900px'
			 * @version 3.45.0
			 */
			ltPropWidth : prop( 'string',{ default : '900px'} ),
            /**
			 * @componentProperty {string} ltPropHeight='500px'
			 * @version 3.45.0
			 */
			ltPropHeight : prop( 'string',{ default : '500px'} ),
            /**
			 * @componentProperty {string} ltPropWidth='900px'
			 * @version 3.45.0
			 */
			ltPropLoop : prop( 'boolean'),
             /**
			 * @componentProperty {boolean} ltPropLoop
			 * @version 3.45.0
			 */
            ltPropPreLoad : prop( 'string', { default : 'metadata'} ),
             /**
			 * @componentProperty {string} ltPropPreLoad='metadata'
			 * @version 3.45.0
			 */
            ltPropAutoPlay : prop( 'boolean' ),
             /**
			 * @componentProperty {boolean} ltPropAutoPlay
			 * @version 3.45.0
			 */
            ltPropPoster : prop( 'string' ),
             /**
			 * @componentProperty {string} ltPropPoster
			 * @version 3.45.0
			 */
            ltPropCaption : prop( 'string', { default : 'off'} ),
             /**
			 * @componentProperty {string} ltPropCaption='off'
			 * @version 3.45.0
			 */
            ltPropQuality : prop( 'string' , { default : 'auto'} ),
             /**
			 * @componentProperty {string} ltPropQuality='auto'
			 * @version 3.45.0
			 */
            ltPropOptions : prop( 'object', { default : {}}),
             /**
			 * @componentProperty {object} ltPropOptions
             * @default {}
			 * @version 3.71.0
			 */
            ltPropPrefetch :  prop( 'boolean', { default : false } ),
            /**
			 * @componentProperty {boolean} ltPropPrefetch=false
			 * @version 3.71.0
			 */
            ltPropPrefetchOptions : prop( 'object', { default : {
                method:'GET',
                mode:'cors'
                }
            }),
            /**
             * @componentProperty {object} ltPropPrefetchOptions
             * @default {"method":"GET","mode":"cors"}
             * @version 3.71.0
             */ 
            ltPropFocus :  prop( 'boolean', { default : false } ),
            /**
			 * @componentProperty {boolean} ltPropFocus=false
			 * @version 3.88.1
			 */
            ltPropMiniMode :  prop( 'boolean', { default : false } ),
            // /**
			//  * @componentProperty {boolean} ltPropMiniMode=false
			//  * @version 3.91.1
			//  */
            ltPropMiniModeOptions :  prop( 'object', { default : {} } ),
            // /**
			//  * @componentProperty {object} ltPropMiniModeOptions
			//  * @version 3.91.1
			//  */
			//system data
            isNotFirefox : prop( 'boolean', { default :  
                window._lyteUiUtils.getBrowser() != "firefox" } ),
			volume :  prop( 'number', { default : 0 }),
			selectedOption : prop( 'string', { default : 'settings'}),
			elapsedTime : prop( 'number', { default : 0}),
			loadedTime : prop( 'number', { default : 0}),
			toolTip : prop( 'object', { default : { 'name' : '', 'time' : '00:00' }}),
			currentDuration : prop( 'string', { default : '00:00'}),
			duration : prop( 'string', { default : '00:00'}),
			subMenuOpened : prop( 'boolean', { default : false }),
			speedData : prop( 'array', { default : [
				{ "label" : "0.25" },
				{ "label" : "0.5" },
				{ "label" : "1" },
				{ "label" : "1.25" },
				{ "label" : "1.5" },
				{ "label" : "1.75" },
				{ "label" : "2" }
			]}),
			captionsData : prop( 'array', { default : [
				{ "label" : "off"}
			]}),
			qualityData : prop( 'array', { default : [
				{ "label" : "auto" }
			]}),
			chaptersData : prop( 'array' ),
            subtitles : prop( 'array'),
            subText : prop( 'string', { default : ''}),
            selectedSource :  prop( 'number', { default : 0 } ),
            renderPrefetch : prop( 'boolean', { default : false } ),
			prefetchLoading : prop( 'string', { default : '' } ),
            metaSetted : prop( 'boolean', { default : false } ),
            durationSec : prop( 'number', { default : 0 }),
            videoStatus : prop( 'string', { default : window._lyteUiUtils.i18n( "lyte.video.play" ) } ),
            volumeStatus : prop( 'string', { default : window._lyteUiUtils.i18n( "lyte.video.volume.mute" ) } ),
            forwardStatus : prop( 'string', { default : '' } ),
            backwardStatus : prop( 'string', { default : 'lyteVideoIconDisabled'} ),
            canPlay : prop( 'boolean', { default : false })
		};
	}

    init() {
		// this.setData('volume',this.getData('ltPropVolume'));
	}

    didConnect() {
        if( !this.data.ltPropPrefetch ){
            this.setGlobals();
        }else{
            this.setData( "renderPrefetch", true );
        }

		this._timerIdx;
		this._menuIdx = 0;
        this._focused =  this.data.ltPropFocus || false;
        this._subIdx = 0;

		this.$node.screenShot = this.screenShot.bind( this );
        this.$node.play = this.play.bind( this );
        this.$node.pause = this.pause.bind( this );
        const poster = this.data.ltPropPoster;
        if( poster && isNaN( poster ) ){
            this.getImage( poster )
        }else{
            this.drawImage( undefined, 'black' );
        }
	}

    setGlobals() {
        this._video = this.$node.querySelector( "video" );
		this._animationWrapper = this.$node.querySelectorAll('.lyteVideoAnimateWrapper');
		this._controls = this.$node.querySelector('.lyteVideoControls');
		this._menu = this.$node.querySelector('.lyteVideoSettingsWrapper');
		this._handler = this.$node.querySelector('.lyteVideoProgressHandler');
		this._playIcon = this.$node.querySelector('.lyteVideoPlayIcon');
        this._container = this.$node.querySelector( '.lyteVideoContainer');
        if( this._menu ){
            this._menu.style.right = "70px";
        }
    }

    didDestroy() {
		// document.body.removeEventListener('keydown', this._keyDown);
		if( this.__preloaded ){
			var src = this.data.ltPropSource;
            for( var i = 0;i < src.length; i++ ){
                window.URL.revokeObjectURL( src[ i ].src );
            }
		}
		delete this._video;
		delete this._playIcon;
		delete this._handler;
		delete this._controls;
		delete this._menu;
		delete this._progressBar;
		delete this._menuListener;
		// delete this._keyDown;
        delete this._focused;
        delete this._menuIdx;
        delete this._controls;
	}

    formatTime(timeInSeconds) {
        if( timeInSeconds ){
		    var result = new Date(timeInSeconds * 1000).toISOString().substr(11, 8);
            window.time = result.substr( 0, 2) != '00' ? result.substr( 0, 2) + ':' + result.substr(3, 2) + ':' + result.substr(6, 2) : result.substr(3, 2) + ':' + result.substr(6, 2);
            return time;
        }else{
            return "00:00";
        }
	}

    convertToSeconds(time) {
        var sec = 0;
        sec = parseInt( time.substr( 0, 2) * 3600 ) + parseInt( time.substr( 3, 2) * 60 ) + parseInt( time.substr( 6, 2) );

        return sec; 
    }

    screenShot() {
        var that = this,
        canvas = document.createElement('canvas'),
        ctx = canvas.getContext('2d'),
        video = this._video;

        ctx.canvas.width = video.videoWidth;
        ctx.canvas.height = video.videoHeight;

        window.createImageBitmap(video).then(function(frame) {
            ctx.drawImage(frame, 0, 0);
        }).then(  function(){
            var url =  canvas.toDataURL();
            that.getMethods('onScreenShot') && that.executeMethod( 'onScreenShot', url, that.$node); 
        });
    }

    classHandler(node, action, className) {
        if( node ){
            node.classList[ action ]( className );
        }
    }

    setDefault() {
        let video = this._video,
        playbackRate = this.getData('ltPropPlayRate'),
        startTime = this.getData( 'ltPropCurrentTime'),
        loop = this.getData( 'ltPropLoop' ),
        autoPlay = this.getData( 'ltPropAutoPlay'),
        chapters = this.getData( 'ltPropChapters'),
        poster = this.getData( 'ltPropPoster' ),
        sources = this.getData( 'ltPropSource' ),
        prefetch = this.getData( 'ltPropPrefetch'),
        muted = this.getData( 'ltPropMuted' ),
        flag = sources[ 0 ].src.includes( "blob:" ),
        miniModeOptions = {
            "resize" : false,
            "draggable" : false,
            "miniWindowWidth" : "400px",
            "miniWindowHeight" : "225px",
            "resizeMinWidth" : "200px",
            "resizeMaxWidth" : video.videoWidth + "px",
            "resizeMinHeight" : "200px",
            "resizeMaxHeight" : video.videoHeight + "px"
        }
        this.setData( "miniModeOptions", Object.assign( miniModeOptions, this.data.ltPropMiniModeOptions ) );
        this.setData('volume', this.data.ltPropVolume);
        video.oncontextmenu = function(){ return false };
        video.onended = function(){
            this.getMethods('onEnded') && this.executeMethod( 'onEnded', this.$node ); 
        }.bind( this );

        if( muted ){
            this._video.muted = true;
        }
        if( !window.navigator.maxTouchPoints ){
            var tooltip = this.$node.querySelector(".lyteVideoToolTip");
            this.classHandler( tooltip, "add", "lyteVideoToolTipHover" );
        }
        if( poster && !isNaN( poster ) ){;
            this.drawImage( poster, true );
        }
        
        
        if( prefetch && !flag ){
            this.setData( "renderPrefetch", true );
            // this.prefetch();
            return;
        }

        if( startTime ){
            video.currentTime = startTime;
        }

        if( autoPlay ){
            video.setAttribute( 'autoplay', '');
        }

        if( playbackRate ){
            video.playbackRate = playbackRate;
        }


        if( loop ){
            video.setAttribute( 'loop', '');
        }

        if( chapters ){
            window.chaptersData = [],
            window.totalLoadedTime = video.buffered.length ? video.buffered.end( video.buffered.length - 1 ) : 0;

            for( window.i = 0; i < chapters.length; i++){
                var width = ( ( this.convertToSeconds( chapters[ i ].endTime ) - this.convertToSeconds( chapters[ i ].startTime )  ) / video.duration * 100 ).toFixed(2),
                loadedTime;

                if( totalLoadedTime > chapters[ i ].endTime ){
                    loadedTime = 100;
                }else if( totalLoadedTime <  chapters[ i ].startTime ){
                    loadedTime = 0;
                }
                else{  
                    loadedTime =  ( totalLoadedTime - chapters[i].startTime ) / (chapters[i].endTime - chapters[i].startTime )* 100;
                    loadedTime = loadedTime > 100 ? 100 : loadedTime;                       
                }

                chaptersData.push( { title :  chapters[ i ].title, startTime : this.convertToSeconds( chapters[ i ].startTime ) , endTime : this.convertToSeconds( chapters[ i ].endTime ), width : width, time : 0, loadedTime : loadedTime } )    
            }
            if( chaptersData.length > 0 ){
                this.classHandler( this.$node.querySelector(".lyteVideoProgressBar"), "add", "lyteVideoBgTransparent" );
            }
            this.setData('chaptersData',chaptersData)
        }
        
        if( this._menu ){
            this._menu.style.right = "70px";
        }
        this.setData( "ltPropCurrentTime", -1 );
    }

    menuListener(evt, fl) {
        if( !this._menu ){
            return
        }
        var flag = true,
        classList = ['lyteVideoSettingsWrapper', 'lyteVideoSettingsItem','lyteVideoSettingsIcon', 'lyteVideoSettingsLabel','lyteVideoSettingsKey','lyteVideoSettings', 'lyteVideoSettingsDropdownHead', 'lyteVideoSettingsBackIcon' ],
        menu = this._menu;

        if( evt ){
            for( window.i = 0; i < classList.length; i++){
                if( evt.target.classList.contains( classList[ i ] )){
                    flag = false;
                }
            }
        }
        if( flag  || fl ) {
            this.classHandler( menu, "add", "lyteVideoMenuHide" );
            this.classHandler( menu, "add", "lyteVideoSettingsItemHover" );
            menu.style = ''
            // menu.style.bottom = '70px';
            menu.style.right = "70px";
            this.setData({'subMenuOpened' : false, 'selectedOption' : 'settings'});
            this._menuIdx = 0;
            document.removeEventListener('scroll',this._onScroll);
            document.removeEventListener('click', this._menuListener);
            window.removeEventListener('resize',this._onResize);
            delete this._menuListener, this._onScroll, this._onResize;
        }
    }

    onScroll(evt) {
        if( this._menu ){
            return;
        }
        var menu = this._menu,
        bcr = menu.getBoundingClientRect();
        if( bcr.top < 0){
            menu.style.bottom = null;
            menu.style.top = this.$node.clientHeight + 'px';
        }else if(bcr.top - 70 > bcr.height ){
            menu.style.bottom = '70px';
            menu.style.top = null;
        }
    }

    onResize() {
        this.onScroll();
    }

    left() {
        return window._lyteUiUtils.getRTL() ? "right" : "left";
    }

    bind_evt(fn, isTch) {
        document[ fn ]( isTch ? 'touchmove' : 'mousemove', this._move, true );
        document[ fn ]( isTch ? 'touchend' : 'mouseup', this._up, true );
    }

    updateTime(evt, flag) {

        var ele = this._controls.querySelector(".lyteVideoProgressBar"),
        video = this._video,
        bcr = ele.getBoundingClientRect(),
        width = bcr.width,
        x = Math.min( Math.max( bcr.left, evt.clientX ), bcr.right ),
        diff = Math.min( Math.abs( x - bcr.left ), width ),
		duration = this.data.durationSec,
        time = Math.max( 0, Math.min( duration - 0.9, ( duration * ( diff / width ) ).toFixed( 2 ) ) ),
        hanlder = this._handler;
        
        if( flag || this.data.renderPrefetch ){
            return time;
        }


        video.currentTime = time;
        
        if( this._mouseDown ){
           var left = ( time / duration * 100) ;
           left = left > 100 ? 100 : left ;
           hanlder.style.left = left + '%';
        }

    }

    progressMouseMove(evt) {
        var touches = evt.touches || [],
		length = touches.length,
		ev = touches[ 0 ] || evt;

		if( length > 1 ){
			return;
		}
        
        if( length ){
            evt.preventDefault();
        }

        var time  = this.updateTime( ev );
        if( this.data.renderPrefetch ){
            this._isPaused = true;
            this._time = time;
            this.prefetch();
        }
    }

    progressMouseUp(evt) {
        var isTch = evt.touches ? true : false;

        this.bind_evt( 'removeEventListener', isTch );
        if( this._paused ){
            this.play();
        }

        delete this._move;
		delete this._up;
        delete this._paused;
        delete this._mouseDown;
    }

    play() {
        if( this._video == undefined){
            return
        }

        var video = this._video,
        _this = this,
        fn = function(){
            delete _this._happening;
            var final = _this._final;

            if( final ){
                delete _this._final;
                _this[ final ]();
            }
        };
        this.setData( "hidePoster", true );

        if( this._happening ){
            this._final = 'play';
        }else if( video.paused ){
            this._happening = true;
            video.play().then( fn ).catch( fn );
        }
    }

    pause() {
        if( this._video == undefined){
            return
        }
        var video = this._video;

        if( this._happening ){
			this._final = 'pause';
		} else {
			video.pause();
		}
    }

    getImage(url) {
        var img = new window.Image(),
        canvas = this.$node.querySelector( '.lyteVideoPoster' ),
        ctx = canvas.getContext( "2d" ),
        video = this._video,
        that = this;
        
        ctx.canvas.width = that.$node.clientWidth;
        ctx.canvas.height = that.$node.clientHeight;
        
        img.onload = function( ){
            ctx.drawImage( img , 0, 0, that.$node.clientWidth, that.$node.clientHeight);
        }

        img.src = url;
    }

    drawImage(time, fl) {
        var that = this,
        canvas =  fl ? this.$node.querySelector( '.lyteVideoPoster' ) : '',
        ctx = canvas.getContext( '2d' );
        
        if( fl == "black" ){
            ctx.fillStyle = "black";
            ctx.fillRect(0, 0, canvas.width, canvas.height);
            return;
        }
        var video = this._video ?  this._video.cloneNode( true ) : '';
        video.currentTime = time;

        video.oncanplaythrough =  function(){
            ctx.canvas.width = that.$node.clientWidth;
            ctx.canvas.height = that.$node.clientHeight;


            window.createImageBitmap(video).then(function(frame) {
                ctx.drawImage(frame, 0, 0, that.$node.clientWidth, that.$node.clientHeight );
            }).catch( function( error ){
                // console.log( error );
            });
        }
    }

    prefetch() {
        var lyteSelf = this;
        var __fetch = window.fetch,
		__this = this,
		__data = __this.data,
		ns = 'prefetchLoading',
		__ns = 'lyteVideoPrefetch',
		renderPrefetch = 'renderPrefetch',
		__error = function( err ){
			__this.setData( ns, __ns + "Error" );

			var cb = "onPrefetchError";
			
			if( __this.getMethods( cb ) && __this.executeMethod( cb, err, __this.$node ) == false ){
				return;
			}

			if( /Access\-Control\-Allow\-Origin/i.test( err.message || "" ) ){
				this.setData( renderPrefetch, false );
			}

		},
		src = __data.ltPropSource[ __data.selectedSource ] || {};

        if( __this.__preloaded ){
			return;
		}

        if( __fetch ){

			var cb = "onBeforePrefetch";

			if( __this.getMethods( cb ) && __this.executeMethod( cb, __this.$node ) == false ){
				return;
			}

			__this.setData( ns, __ns + 'Loading' );
            __this.setData( "hidePoster", false );
			__fetch( src.src || "", __data.ltPropPrefetchOptions ).then( function( res ){
				if( /^2..$/.test( res.status ) ){
					return res.blob();
				} else {
					__error( {} );
				}
			}).then( function( blob ){
				
				__this.setData( ns, __ns + 'Success' );

				var __url = window.URL.createObjectURL( blob ),
				cb = "onPrefetchSuccess";
				
				lyteSelf.$addon.objectUtils( src, 'add', 'src', __url );
				__this.getMethods( cb ) && __this.executeMethod( cb, __url, __this.$node );
				__this.setData( renderPrefetch, !( __this.__preloaded = true ) );
                __this.setGlobals();
			}).then( function(){

                if( !__this._isPaused ){
                    __this.play();
                }
                if( __this._time ){
                    __this.setData( "hidePoster", true );
                    __this._video.currentTime = __this._time;
                }
                if( __this._curTime ){
                    __this._video.currentTime = __this._curTime;
                    __this.setData( "hidePoster", true );
                }
                if( __this._volume ){
                    __this._video.volume = __this._volume;
                    __this.setData('volume',__this._volume );
                }
                if( __this._muted ){
                    __this._video.muted = __this._muted;
                }
                delete __this._isPaused;
                delete __this._time;
                delete __this._curTime;
                delete __this._volume;
                delete __this._muted;
            }).catch( __error );
		}
    }

    handlePreVolume(currentValue) {
        var volIcon = this.$node.querySelector(".lyteVideoVolumeIcon");
        if( currentValue == "mute" ){
            this._muted = !this._muted;
            this._volume = this._volume == undefined ? 1 : this._volume; 
            if( this._muted ){
                this.classHandler( volIcon, "add", "lyteVideoMuted" );
                this.setData('volume', 0);
            }else{
                this.classHandler( volIcon, "remove", "lyteVideoMuted" );
                this.setData('volume', this._volume );
            }
            return;
        }
        if( this.data.renderPrefetch ){
            this._volume = currentValue.value;
            if( currentValue.value == 0 ){
                this.classHandler( volIcon, "add", "lyteVideoMuted" );
            }else if( currentValue.value < 0.5) {
                this.classHandler( volIcon, "remove", "lyteVideoMuted" );
                this.classHandler( volIcon, "add", "lyteVideoLow" );
            }else{
                this.classHandler( volIcon, "remove", "lyteVideoLow" );
            }
        }
    }

    static actions() {
        return {
            keyDown : function( evt ){
                if( !this._video ){
                    return
                }
                var video = this._video,
                wrapperEle,
                i = 0;
                if( this._focused ){  
                    switch( evt.keyCode ){
                        case 32 : 
                            evt.preventDefault();
                            if( this.data.renderPrefetch ){
                                this.prefetch();
                                return;
                            }
                            wrapperEle = this._animationWrapper[1];
                            window.iconEle = wrapperEle.children[0];
                            if( video.paused ){
                                this.classHandler( iconEle, "remove", "lyteVideoAnimatePauseIcon" );
                                this.classHandler( iconEle, "add", "lyteVideoAnimatePlayIcon" );
                                this.play();
                            }else{
                                this.classHandler( iconEle, "add", "lyteVideoAnimatePauseIcon" );
                                this.classHandler( iconEle, "remove", "lyteVideoAnimatePlayIcon" );
                                this.pause();
                            }
                            break;
                        case 39 :
                            evt.preventDefault();
                            wrapperEle = this._animationWrapper[2];
                            video.currentTime += 10;
                            break;
                        case 37 :
                            evt.preventDefault();
                            wrapperEle = this._animationWrapper[0];
                            video.currentTime -= 10;
                            break;
                        case 38 :
                            var menu = this._menu;
                            if( menu && !menu.classList.contains('lyteVideoMenuHide')){
                                this.classHandler( menu, "remove", "lyteVideoSettingsItemHover" );
                                evt.preventDefault();
                                var options = menu.children;
                                this._menuIdx = this._menuIdx == 0 ? options.length  : this._menuIdx;
                                
                                window.prev =  Math.abs( this._menuIdx % options.length ),
                                window.idx =  Math.abs( --this._menuIdx % options.length );
        
                                this.classHandler( options[ prev ], "remove", window.lyteVideoSettingsItemSelected );
                                menu.scrollTop =  options[ idx ].offsetTop;
                                this.classHandler( options[ idx ], "add", window.lyteVideoSettingsItemSelected );
                            }
                            break;
                        case 40 :
                            var menu = this._menu;
                            if( menu && !menu.classList.contains('lyteVideoMenuHide')){
                                evt.preventDefault();
                                this.classHandler( menu, "remove", "lyteVideoSettingsItemHover" );
                                var options = menu.children,
                                prev = Math.abs( this._menuIdx % options.length ),
                                idx = Math.abs( ++this._menuIdx % options.length );
        
                                this.classHandler( options[ prev ], "remove", window.lyteVideoSettingsItemSelected );
                                menu.scrollTop =  options[ idx ].offsetTop;
                                // options[ idx ].scrollIntoView({ block: 'nearest'});
                                this.classHandler( options[ idx ], "add", window.lyteVideoSettingsItemSelected );
        
                            }
                            break;
                        case 13 :
                            var menu = this._menu;
                            if( menu && !menu.classList.contains('lyteVideoMenuHide')){
                                // evt.preventDefault();
                                var options = menu.children,
                                idx = Math.abs( this._menuIdx % options.length );
                                
                                options[ idx ].click();
                            }
                            evt.preventDefault();
                            break;
                    }
                    if( wrapperEle ){
                        wrapperEle.classList.add('lyteVideoAnimate');
                        setTimeout( function(){ wrapperEle.classList.remove('lyteVideoAnimate')}.bind(this),500);
                    }
                }
            },
            play : function( evt ){
                if( !this.data.hidePoster ){
                    this.setData( "hidePoster", true );
                }
                this.classHandler( this._playIcon, "remove", "lyteVideoPaused" );
                this.setData( "videoStatus",  window._lyteUiUtils.i18n( "lyte.video.pause" ) );
                this.getMethods('onPlay') && this.executeMethod( 'onPlay', this._video, evt, this.$node);
            },

            pause : function( evt ){
                this.classHandler( this._playIcon, "add", "lyteVideoPaused" );
                this.setData( "videoStatus",  window._lyteUiUtils.i18n( "lyte.video.play" ) );
                this.getMethods('onPause') && this.executeMethod( 'onPause', this._video, evt, this.$node);
            },

            togglePlayPause : function( ){
                var video = this._video,
                fn = "pause";

                if( video.paused ){
                    fn = "play"
                }

                this[ fn ]();
                
            },

            prefetch : function(){
                if( this.data.ltPropPrefetch ){
                    this.setData( "renderPrefetch", true );
                    this.prefetch();
                }
            },

            clickOnVideo : function(){
                if( this.data.renderPrefetch ){
                    this.prefetch();
                    return;
                }
                
                var video = this._video,
                iconEle = this._animationWrapper[1].children[0];
                if( document.pictureInPictureElement == video){
                    document.exitPictureInPicture();
                }else if( video.paused ){
                    this.classHandler( iconEle, "remove", "lyteVideoAnimatePauseIcon" );
                    this.classHandler( iconEle, "add", "lyteVideoAnimatePlayIcon" );
                    this.play();
                }else{
                    this.classHandler( iconEle, "add", "lyteVideoAnimatePauseIcon" );
                    this.classHandler( iconEle, "remove", "lyteVideoAnimatePlayIcon" );
                    this.pause();
                }
                this.classHandler( this._animationWrapper[ 1 ], "add", "lyteVideoAnimate" );
                setTimeout( function(){ this.classHandler( this._animationWrapper[ 1 ], "remove", "lyteVideoAnimate" );}.bind(this),500);
            },

            skipVideo : function( type ){
                var video = this._video,
                currrentTime = video.currentTime;
                if( type === 'forward' && this.data.forwardStatus != "" ){
                    return;
                }else if( type === 'rewind' && this.data.backwardStatus != '' ){
                    return
                }
                video.currentTime = currrentTime + ( type == 'forward' ? 10 : -10 );
            },

            mute : function(){
                var video = this._video;
                if( this.data.renderPrefetch ){
                    this.handlePreVolume( "mute" );
                    return;
                }
                video.muted = !video.muted;
            },

            changeVolume : function(){
                if( this._video == undefined){
                    return
                }
                var video = this._video,
                volIcon = this.$node.querySelector(".lyteVideoVolumeIcon");

                if(video.volume < 0.5) {
                    this.classHandler( volIcon, "add", "lyteVideoLow" );
                }else{
                    this.classHandler( volIcon, "remove", "lyteVideoLow" );
                }

                if( video.volume > 0 && volIcon && volIcon.classList.contains('lyteVideoMuted')){
                    this.classHandler( volIcon, "remove", "lyteVideoMuted" );
                    video.muted = false;
                    this.setData( { 
                        "volume" : video.volume,
                        "volumeStatus" : window._lyteUiUtils.i18n( "lyte.video.volume.mute" )
                    });
                }else if( !video.muted && video.volume == 0){
                    this.setData({"volume" : 1, 'ltPropVolume' : 1});
                }else if( video.muted ){
                    this.classHandler( volIcon, "add", "lyteVideoMuted" );
                    this.classHandler( volIcon, "remove", "lyteVideoLow" );
                    this.setData( { 
                        'volume' : 0,
                        'volumeStatus' : window._lyteUiUtils.i18n( "lyte.video.volume.unmute" )
                    });
                }
            },

            update : function( evt ){
                if( this._video == undefined){
                    return
                }
                var video = this._video,
                time = video.currentTime,
                duration = video.duration,
                chaptersData = this.getData( 'chaptersData' ),
                subtitles = this.getData( 'subtitles' ),
                subEle = this.$node.querySelector( '.lyteVideoSubtitle'),
                curTime;
                
                
                if( subtitles ){
                    for( window.i = 0; i < subtitles.length; i++){
                        if( time >= subtitles[ i ].startTime && time <= subtitles[ i ].endTime && subEle.innerHTML == ''){
                            this.setData('subText', subtitles[ i ].text);
                            this._subIdx = i;
                        }
                    }
                    if( time <= subtitles[ this._subIdx ].startTime  || time >= subtitles[ this._subIdx ].endTime) {
                        this.setData('subText', '');
                    }
                }

                if( chaptersData ){
                    for( window.i = 0; i < chaptersData.length; i++){
                        if( time > chaptersData[i].endTime){
                            curTime = 100;
                        }else if( time <  chaptersData[i].startTime ){
                            curTime = 0;
                        }
                        else{  
                            curTime =  ( time - chaptersData[i].startTime) / (chaptersData[i].endTime - chaptersData[i].startTime )* 100;

                            curTime = curTime > 100 ? 100 : curTime;
                            this._curChapter = chaptersData[i].title;
                        }     
                        this.$addon.objectUtils( chaptersData[i] , "add" , "time", curTime );
                    }
                }
                
                this.setData( { 'currentDuration' : this.formatTime( time ), 'elapsedTime' : ( time / duration * 100) });
                this._handler.style.left = ( time / duration * 100) + '%';
                if( time > 10 ){
                    this.setData( "backwardStatus", "" );
                }else{
                    this.setData( "backwardStatus", "lyteVideoIconDisabled" );
                }
                if( ( duration - time ) < 10 ){
                    this.setData( "forwardStatus", "lyteVideoIconDisabled" );
                }else{
                    this.setData( "forwardStatus", "" );
                } 

                this.getMethods('onProgress') && this.executeMethod( 'onProgress', this._video, time, evt, this.$node);               
            },

            meta : function(){
                if( this._video == undefined){
                    return
                }

                var video = this._video,
                duration = video.duration;
                if( !this.data.metaSetted )
                {
                    this.setData({ 
                        'metaSetted' :  true ,
                        'duration' : this.formatTime( duration ),
                        'durationSec' : duration
                    })
                    this.setDefault();
                }
                this.getMethods('onMetaLoaded') && this.executeMethod( 'onMetaLoaded', this.$node, {
                    height : video.videoHeight,
                    width : video.videoWidth,
                    duration : video.duration
                } );
            },

            togglePip : function(){
                if (!('pictureInPictureEnabled' in document)) {
                    console.warn("Picture-in-picture is not supported");
                    return;
                }
                var video = this._video;
                if( video !== document.pictureInPictureElement ){

                    video.requestPictureInPicture().then( function(){
                        this.classHandler( this._controls, "remove", "lyteVideoControlsShow" );
                    }.bind(this)).catch(function(error){ 
                        console.log(error);
                    });
                }else{
                    document.exitPictureInPicture();
                }

            },

            toggleMiniMode : function( flag ){
                const miniContainer = this.$node.querySelector( ".lyteVideoMiniContainer" );
                const miniMode = this.data.miniMode;
                if( miniMode ){
                    window._lyteUiUtils.appendChild( this._container, this._video );
                    window._lyteUiUtils.appendChild( this._container, this._controls );
                }else{
                    window._lyteUiUtils.appendChild( miniContainer, this._video );
                    window._lyteUiUtils.appendChild( miniContainer, this._controls );
                }
                if( flag ){
                    this.pause();
                }
                this.setData( "miniMode", !this.data.miniMode );
            },
            toggleFullScreen : function( ele ){
                var videoContainer = this.$node.querySelector('.lyteVideoContainer'),
                ele = ele ? ele : this.$node.querySelector('.lyteVideoFullScreen'),
                options = this.data.ltPropOptions,
                that = this;
                if( options && options.fullScreen === false ){
                    return
                }
                if( document.fullscreenElement || document.webkitFullscreenElement){               
                    document.fullscreenElement ? document.exitFullscreen() : document.webkitExitFullscreen();
                    that.classHandler( ele, "remove", "lyteVideoFullScreenExit" );
                }else{
                    var fn = videoContainer.requestFullscreen ? 'requestFullscreen' : 'webkitRequestFullscreen' ;
                    videoContainer[ fn ]().then( function(){
                        this.classHandler( ele, "add", "lyteVideoFullScreenExit" );
                        if( window.navigator.maxTouchPoints ){
                            window.screen.orientation.lock('landscape').catch( function( error ){ });
                        }
                    }).catch( function(error){
                        console.log( error)
                    });
                   
                }
            },

            toggleSettings : function(){
                var menu = this._menu,
                diff,bcr;

                if( menu && menu.classList.contains('lyteVideoMenuHide')){
                    
                    menu.classList.remove('lyteVideoMenuHide');

                    this._menuListener = this.menuListener.bind( this );

                    document.addEventListener('click', this._menuListener);

                }else{
                    this.menuListener(null, true);
                }
            },

            mouseMove : function( evt, ele ){
                var container = this.data.miniMode ? ele : this._container;
                if( evt.pointerType != 'touch' && document.pictureInPictureElement == null ){
                    clearTimeout( this._timerIdx );
                    this.classHandler( container , "add", "lyteVideoControlsShow" );
                    var flag = this._menu ? this._menu.classList.contains('lyteVideoMenuHide') : true;
                    if( flag  &&  this._video && !this._video.paused){
                        this._timerIdx =  setTimeout(  function(){ 
                            this.classHandler( container , "remove", "lyteVideoControlsShow" );
                        }.bind(this), 3000 );
                    }
                }
            },

            mouseEnter : function( evt, ele ){
                this._focused = true;
                var container = this.data.miniMode ? ele : this._container,
                flag = this._menu ? this._menu.classList.contains('lyteVideoMenuHide') : true;
                if( document.pictureInPictureElement == null ){
                    this.classHandler( container, "add", "lyteVideoControlsShow" );
                }
                if( evt.pointerType == 'touch' && flag){
                    clearTimeout( this._timerIdx );
                    this._timerIdx =  setTimeout(  function(){ 
                        this.classHandler( container, "remove", "lyteVideoControlsShow" );
                    }.bind(this), 3000 );
                }
            },

            mouseLeave : function( evt, ele ){
                this._focused = false;
                var container = this.data.miniMode ? ele : this._container,
                flag = this._menu ? this._menu.classList.contains('lyteVideoMenuHide') : true;
                if( evt.pointerType != 'touch' && flag && this._video && !this._video.paused){
                    this.classHandler( container , "remove", "lyteVideoControlsShow" );
                }
            },

            mouseOver : function(){
                var menu = this._menu;
                if( menu && !menu.classList.contains('lyteVideoSettingsItemHover')){
                    var options = menu.children,
                    idx = Math.abs( this._menuIdx % options.length);
                    this.classHandler( options[ idx ], "remove", "lyteVideoSettingsItemSelected" );
                    this.classHandler( this._menu, "add", "lyteVideoSettingsItemHover" );
                }
            },

            settingsMenuClick : function( ele ){
                var value = ele.getAttribute('data-value'),
                menu = this._menu,
                bck = this.getMethods('onMenuClick') && this.executeMethod( 'onMenuClick', value, menu );
                bck = ( bck == undefined ) ? true : bck;
                
                if( !menu ){
                    return;
                }
                if( !bck ){
                    this.classHandler( menu, "add", "lyteVideoMenuHide" );
                    this.setData({'subMenuOpened' : false, 'selectedOption' : 'settings'});
                    return;
                }

                if( value == "back" ){
                    this.classHandler( menu, "add", "lyteVideoMenuHide" );
                    this.setData({'subMenuOpened' : false, 'selectedOption' : 'settings'});
                    this.classHandler( menu, "remove", "lyteVideoMenuHide" );
                }else if( this.data.selectedOption == "settings"){
                    this.setData({'subMenuOpened' : true, 'selectedOption' : value });
                    this.classHandler( menu, "remove", "lyteVideoMenuHide" );
                }else {
                    if( this.data.selectedOption === "playBackSpeed" ){
                        this.setData( 'ltPropPlayRate', value );
                    }else if( this.data.selectedOption === "qualityOption" ){
                        this.setData('ltPropQuality', value );
                    }else if( this.data.selectedOption === "subtitles" ){
                        this.setData( 'ltPropCaption', value );
                    }
                    this.classHandler( menu, "add", "lyteVideoMenuHide" );
                    this.setData({'subMenuOpened' : false, 'selectedOption' : 'settings'});
                }
            },

            progress : function( evt ){
                if( this._video == undefined){
                    return
                }
                var video = this._video,
                buffer = video.buffered,
                duration = video.duration,
                chaptersData = this.getData( 'chaptersData' ),
                totalLoadedTime, width, loadedTime;

                if( buffer.length ){ 
                    totalLoadedTime = buffer.end( buffer.length - 1 );    
                    width = totalLoadedTime / duration * 100;
                } 

                if( chaptersData  && totalLoadedTime ){

                    for( window.i = 0; i < chaptersData.length; i++){
                        if( totalLoadedTime > chaptersData[i].endTime){
                            loadedTime = 100;
                        }else if( totalLoadedTime <  chaptersData[i].startTime ){
                            loadedTime = 0;
                        }
                        else{  
                            loadedTime =  ( totalLoadedTime - chaptersData[i].startTime ) / (chaptersData[i].endTime - chaptersData[i].startTime )* 100;
                            loadedTime = loadedTime > 100 ? 100 : loadedTime;                       
                        }    

                        this.$addon.objectUtils( chaptersData[i] , "add" , "loadedTime", loadedTime );
                    }
                }

                this.setData( 'loadedTime', width)
            },

            progressClick : function( evt ){
                var ele = evt.target;

                if( ele.classList.contains('.lyteVideoProgressHandler') ){
                    return;
                }

                var time = this.updateTime( evt );
                if( this.data.renderPrefetch ){
                    this._isPaused = true;
                    this._time = time;
                    this.prefetch();
                }
            },

            progressMouseDown : function( evt ){
                var touches = evt.touches || [],
                length = touches.length,
                isTch = length != 0;

                if( length > 1 ){
                    return;
                }

                if( this._video && !this._video.paused){
                    this.pause();
                    this._paused = true;
                }
                this._mouseDown = true;
                this._move = this.progressMouseMove.bind( this );
                this._up = this.progressMouseUp.bind( this );
                this.bind_evt( "addEventListener", isTch );
                evt.preventDefault();
            },

            chapterSelect : function( ele ){
                var video = this._video;
                
                video.currentTime = ele.getAttribute('data-start')
            },

            updateToolTip : function( evt ){
                if( this.data.miniMode ){
                    return;
                }
                var time = this.updateTime( evt, true ),
                curChapter = this._curChapter ? this._curChapter : '',
                tooltip = this.$node.querySelector(".lyteVideoToolTip"),
                title = $L(evt.target).closest('.lyteVideoChapter').attr('data-label'),
                ele = this.$node.querySelector(".lyteVideoProgressBar"),
                bcr = ele.getBoundingClientRect(),
                op = {};
                
                title = title ? title : curChapter;
                tooltip.style.left = evt.clientX - bcr.left -  15  + "px";
                op.name = title;
                op.time = this.formatTime( time );

                this.setData('toolTip', op);
            },
            onCanPlay : function(){
                this.setData("canPlay", true );
                this.getMethods('onCanPlay') && this.executeMethod( 'onCanPlay', this.$node );
            },
            onWaiting : function( isBuffering ){
                this.setData("canPlay", isBuffering );
            }
        };
    }

    static methods() {
        return {
            setVolume : function( handlerIndex , currentValue ){
                if( this.data.renderPrefetch ){
                    this.handlePreVolume( currentValue );
                    return;
                }
                var video = this._video;
                if( this._video == undefined){
                    return
                }
                if( currentValue.value == 0){
                    video.muted = true;
                }
                
                this.setData('ltPropVolume',currentValue.value)
            }
            
        };
    }

    static observers() {
        return {
            playBackRate : function( change ){
                var video = this._video;

                video.playbackRate = change.newValue;
            }.observes( 'ltPropPlayRate' ),

            qualityChange :  function( change ){
                if( !this._video ){
                    return;
                }
                var video = this._video,
                source = video.querySelector('source'),
                curTime = video.currentTime,
                isPaused = video.paused,
                sources = this.data.ltPropSource,
                newValue = change ? change.newValue : this.data.ltPropQuality,
                i;

                if( newValue == "auto" ){
                    return
                }
                for( i = 0; i < sources.length; i++){
                    if( sources[i].size == newValue ){
                        break;
                    }
                }
                this.setData( "selectedSource", i );
                if( this.data.ltPropPrefetch && !sources[ i ].src.includes( "blob:" ) ){
                    this.drawImage( undefined, 'black' );
                    this.setData( "renderPrefetch", true );
                    this._curTime = curTime;
                    this._isPaused = isPaused;
                    this.__preloaded = false;
                    this.prefetch();
                    return;

                }
                for( window.attr in sources[i] ){
                    source.setAttribute(window.attr, sources[i][window.attr] );
                }

                video.load();

                var fn = function(){ 
                    video.currentTime = curTime;
                    if( isPaused ){
                        this.pause();
                    }else{
                        this.play();
                    }
                    video.removeEventListener('canplay',fn);
                }.bind(this);

                video.addEventListener('canplay', fn );

            }.observes( 'ltPropQuality' ).on( 'didConnect' ),

            setSubtitle : function( change ){
                if( !this._video ){
                    return;
                }
                var video = this._video,
                tracks = this.data.ltPropTracks,
                source = video.querySelector('track'),
                capDiv = this.$node.querySelector( '.lyteVideoSubtitleWrapper'),
                subtitles = [],
                newValue = change ? change.newValue : this.data.ltPropCaption,
                track, i;
                
                if( newValue === 'off' && capDiv){
                    this.classHandler( capDiv, "add", "lyteVideoMenuHide" );
                    return;
                }else if( newValue === 'off'){
                    return;
                }
                else if( capDiv ){
                    this.classHandler( capDiv, "remove", "lyteVideoMenuHide" );
                }

                for( i = 0; i < tracks.length; i++){
                    if( tracks[i].label == newValue  || tracks[i].lang == newValue){
                        break;
                    }
                }
                
                for( window.attr in tracks[i] ){
                    source.setAttribute( window.attr, tracks[i][ window.attr ]);
                }
                track = video.textTracks[ 0 ];
                track.mode = "hidden";
                
                if(  tracks && !('src' in tracks[ i ] ) ){
                    subtitles = tracks [ i ].texts.slice();
                    this.setData( 'subtitles', subtitles );
                    return;
                }
                setTimeout( function(){
                    for( window.j = 0; window.j < track.cues.length;window.j++){
                        subtitles.push({ 'startTime' : track.cues[ window.j ].startTime, 'endTime' : track.cues[ window.j ].endTime, 'text' : track.cues[ window.j ].text } )
                    }
                    this.setData( 'subtitles', subtitles );
                }.bind( this ),500);
                
            }.observes( 'ltPropCaption' ).on( 'didConnect' ),

            setVolume : function( change ){
                if( this._video == undefined){
                    return
                }
                var video = this._video,
                slider = this.$node.querySelector('lyte-multislider');
                if( slider.ltProp('value')[0].value != change.newValue ){
                    this.setData('volume',change.newValue );
                }
                if( this.data.renderPrefetch ){
                    this._volume = change.newValue;
                }else{
                    video.volume = change.newValue;
                }
            }.observes( 'ltPropVolume' ),

            loop : function( change ){
                var video = this._video;
                if( change.newValue ){
                    video.setAttribute('loop','');
                }else{
                    video.removeAttribute( 'loop' );
                }
            }.observes( 'ltPropLoop'),

            source : function( change ){
                var video = this._video,
                isPaused = video.paused;
                this.setData( "selectedSource", 0 );
                video.load();
                var fn = function(){ 
                    if( isPaused ){
                        this.pause();
                    }else{
                        this.play();
                    }
                    video.removeEventListener('canplay',fn);
                }.bind(this);

                video.addEventListener('canplay', fn );
            }.observes( 'ltPropSource'),

            time : function( change ){
                var video = this._video;
                if( change.newValue != -1 ){
                    video.currentTime = change.newValue;
                    this.setData( "ltPropCurrentTime", -1 );
                }
            }.observes( 'ltPropCurrentTime'),

            autoPlay : function( change ){
                var video = this._video;

                if( change.newValue ){
                    video.setAttribute( 'autoplay', '');
                }else{
                    video.removeAttribute( 'autoplay');
                }
            }.observes( 'ltPropAutoPlay'),

            muteObserves : function( change ){
                var video = this._video;
                if( this.data.renderPrefetch ){
                    this.handlePreVolume( "mute" );
                    return;
                }
                video.muted = change.newValue;
            }.observes('ltPropMuted'),
            posterObserves : function( change ){
                const poster = change.newValue;
                if( poster && isNaN( poster ) ){
                    this.getImage( poster )
                }else if( poster && !isNaN( poster ) ){
                    this.drawImage( poster, true );
                }else{
                    this.drawImage( undefined, 'black' );
                }
            }.observes( 'ltPropPoster' )
        };
    }
}

LyteUiComponentComponentRegistry.registerHelper( "lyteUiSetSelectedClass" , function( value, selectedVal ){
	if(value == selectedVal){
		return "lyteVideoOptionselected";
	}
	return "";
});
/**
 * @syntax staticBuilder
 * <lyte-video lt-prop-source='[{"src":"https://cdn.plyr.io/static/demo/View_From_A_Blue_Moon_Trailer-576p.mp4","type":"video/mp4","size":"480p"}]'></lyte-video> 
 */
export { LyteVideoComponent };
