// @zoho/lyte-ui-component/components/javascript/view/listview/lyte-list-filter.js
import { prop } from "@slyte/core";
import $L from "@zoho/lyte-dom";
import { Component, objectUtils } from "@slyte/component";

class LyteListFilterComponent extends Component {
    constructor() {
        super();
    }

    init() {
		this.data.ltPropModalProperties != void 0 ? this.setData( 'ltPropModal' , this.data.ltPropModalProperties ) : null;
		
		var data = this.data,
		fields = data.ltPropFields,
		final = [],
		obj = {};

		fields.forEach( function( item ){
			var __data = item.data,
				selected = data.ltPropConditions[item.data.prop];
			if( __data.dataType ){
				final.push({
					data : __data,
					render : window._lyteUiUtils.lyteUiIsEmpty( selected ),
					selected : selected,
					temp : {},
					reset : false,
					class : selected && selected.isValid ? 'lyteListFilterApplied' : ''
				});

				obj[ __data.prop ] = {};
			}
		});
			
		this.$node.reset = function( key ){
			this.reset( key == void 0 ? undefined : this.data.fields.filter( function( item ){
			    return key == item.data.prop   
			}) );
		}.bind( this );

		if( Object.keys(this.data.ltPropConditions).length == 0 ){
			this.setData( "ltPropConditions" , obj );
		}

		this.setData({
			fields : final,
			noResult : fields.length ? 'lyteSearchHidden' : ''
		});

		this.setData('finalStatus', this.data.ltPropAny);
	}

    data() {

		var __array = "array",
		__string = "string",
		__boolean = "boolean";

		return {
			ltPropModal : prop( __string, { default : '{"wrapperClass" : "lyteListModalWrapper", "maxHeight" : "70%", "animation" : "slideFromRight","offset" : { "right" : 0 } }' } ),

			ltPropFields : prop( __array, { default : [] , watch : true } ),
			ltPropContent : prop( __array, { default : [] } ),

			ltPropShow : prop( __boolean, { default : false } ),

			ltPropAny : prop( __boolean ),
			ltPropReset : prop( __boolean, { default : false } ),

			ltPropFormat : prop( __string, { default : "MM-DD-YYYY" } ),
			ltPropTimeFormat : prop( __string, { default : undefined } ),

			ltPropConditions : prop( 'object', { default : {}, watch : true } ),
			ltPropNoResult : prop( __string, { default : window._lyteUiUtils.i18n( 'no.results.found', void 0, 'No Results Found' ) } ),
			ltPropFilteredContent : prop( __array , { default : [] } ),
			ltPropModalProperties : prop( 'object' , { default : undefined } ),
			ltPropFilterElementProperties : prop( 'object', { default : {} } ),

			fields : prop( __array, { default : [] } ),

			finalStatus : prop( __boolean, { default : true } ),
			noResult : prop( __string, { default : '' } )
		};		
	}

    execute(cb) {
		return this.getMethods( cb ) && this.executeMethod.apply( this, arguments );
	}

    static methods() {
        return {

            search : function( arr ){
                this.setData( 'noResult', arr.length ? 'lyteSearchHidden' : '' );
            },

            customReset : function( data, condition, node ){
                return this.execute( 'onCustomFilterReset', data, condition, node );
            },

            fetchMoreData : function( first, data, picklistOptions ){
                return this.execute( 'fetchMoreData', first, data, picklistOptions );
            },

            picklistData : function( value, cell_data, old_value ){
                return this.execute( 'onPicklistConstruct', value, cell_data, old_value );
            },	

            beforeOpen : function(){
                var __elem = arguments[ 1 ],
                index = parseInt( __elem.getAttribute( 'index' ) ),
                __data = this.data.fields[ index ];

                objectUtils( __data, 'add', 'render', true );
                return this.execute( 'onBeforeFilterOpen', __data.data, __data.selected, __elem.querySelector( '.lyteFilterElement' ) );
            },

            reset : function( item ){
                objectUtils( item, 'add', 'value', '' );
            },

            beforeClose : function(){
                var elem = arguments[ 2 ].$node.querySelector( '.lyteAccordionActive .lyteFilterElement' ),
                __data = ( ( elem || {} ).component || {} ).data;

                __data && this.execute( 'onBeforeFilterClose', __data.ltPropData, __data.ltPropCondition, elem );

                if( __data && !__data.ltPropTempCondition.isValid ){
                    elem.ltProp( 'reset', true );
                }
            }
        };
    }

    first_extn(obj) {
		return $L.extend( true, {}, obj );
	}

    reset(fields) {
		var __data = this.data,
		conditions = __data.ltPropConditions,
		Lc = objectUtils,
		__this = this;

		this.setData( 'ltPropFilteredContent' , undefined );

		fields = fields || __data.fields;

		fields.forEach( function( item ){
			var f_data = item.data,
			__prop = f_data.prop;

			Lc( item, 'add', 'reset', true );
			Lc( conditions, 'add', __prop, {} );
		});

		this.setData( 'ltPropAny', __data.finalStatus );
	}

    apply() {
		var __data = this.data,
		conditions = __data.ltPropConditions,
		fields = __data.fields,
		Lc = objectUtils,
		__this = this;

		fields.forEach( function( item ){
			var f_data = item.data,
			extended = __this.first_extn( item.temp );
			extended.filterName = f_data.filterName;

			Lc( conditions , 'add', f_data.prop, extended );
			
			if( conditions[f_data.prop].isValid ){
				Lc( item, 'add', 'class',  'lyteListFilterApplied');
			}else{
				Lc( item, 'add', 'class',  '');
			}
			// Lc( conditions, 'add', f_data.prop, __this.first_extn( item.selected ) );
		});

		this.setData( 'finalStatus', __data.ltPropAny );

		var cb = "onModalFilterApply";
		if( this.getMethods(cb) ){
			setTimeout(function(){
				this.executeMethod( cb , conditions );
			}.bind(this), 0);
		}
	}

    cancel() {
		var __data = this.data,
		conditions = __data.ltPropConditions,
		fields = __data.fields,
		Lc = objectUtils,
		__this = this;

		fields.forEach( function( item ){
			Lc( item, 'add', 'selected', __this.first_extn( conditions[ item.data.prop ] ) );
		});
	}

    callPicklistContruct(index) {
		let data = this.data.fields;

		objectUtils(data[index], 'add', 'reset', true);
	}

    static actions() {
        return {
            reset : function(){
                let ret = true;
                this.reset();
                if( this.getMethods('onReset') ){
                    ret = this.executeMethod('onReset');
                }
                this.setData( 'ltPropShow', !ret );
            },

            apply : function(){
                let ret = true;
                this.apply();
                if( this.getMethods('onApply') ){
                    ret = this.executeMethod('onApply');
                }
                this.setData( 'ltPropShow', !ret );
            },

            cancel : function(){
                let ret = true;
                if( this.getMethods('onCancel') ){
                    ret = this.executeMethod('onCancel');
                }
                this.setData( 'ltPropShow', !ret );
            }
        };
    }

    static observers() {
        return {
            cond_obs : function(){

                var cb = "onConditionChange";

                if( this.getMethods( cb ) ){
                    clearTimeout( this.__timeout );
                    this.__timeout = setTimeout( function(){
                        var __data = this.data;
                        this.executeMethod( cb, __data.ltPropConditions, __data.ltPropAny , true, this.$node );
                    }.bind( this ), 0 );
                }
            }.observes( 'ltPropConditions.*', 'finalStatus' ),

            reset_obs : function( arg ){
                if( arg.newValue ){
                    this.reset();
                    this.setData( arg.item, false );
                }
            }.observes( 'ltPropReset' ),

            show_obs : function( arg ){
                if( !arg.newValue ){
                    this.cancel();
                } 
            }.observes( 'ltPropShow' )
        };
    }
}

export { LyteListFilterComponent };
